// @arcgis/map-components/dist/esm/arcgis-utility-network-validate-topology.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-2198b48c.js';
import { a1 as newWidgetsUtilityNetworkValidateTopology } from './index-08887307.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils-97b1c3e3.js';
import { o as onDisconnectedHelper } from './disconnected-45c113cb.js';

const utilityNetworkValidateTopologyCss = "arcgis-utility-network-validate-topology{}";

const ArcgisUtilityNetworkValidateTopology = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisUtilityNetworkValidateTopologyReady = createEvent(this, "arcgisUtilityNetworkValidateTopologyReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.extentToValidate = "current";
        this.utilityNetwork = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    extentToValidateWatcher(value) {
        if (this.widget) {
            this.widget.extentToValidate = value;
        }
    }
    utilityNetworkWatcher(value) {
        if (this.widget) {
            this.widget.utilityNetwork = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsUtilityNetworkValidateTopology(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisUtilityNetworkValidateTopologyReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        if (this.extentToValidate !== undefined) {
            params.extentToValidate = this.extentToValidate;
        }
        if (this.utilityNetwork !== undefined) {
            params.utilityNetwork = this.utilityNetwork;
        }
        return params;
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "extentToValidate": ["extentToValidateWatcher"],
        "utilityNetwork": ["utilityNetworkWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
};
ArcgisUtilityNetworkValidateTopology.style = utilityNetworkValidateTopologyCss;

export { ArcgisUtilityNetworkValidateTopology as arcgis_utility_network_validate_topology };
