// @arcgis/map-components/dist/esm/arcgis-version-management-version-item_2.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-2198b48c.js';
import { i as importCoreReactiveUtils } from './index-08887307.js';

const ArcgisVersionManagementVersionItem = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.versionItemActionClickedEvent = createEvent(this, "versionItemActionClickedEvent", 7);
        this.strings = undefined;
        this.hasAdvEditingUte = undefined;
        this.isVersioningApiAvailable = undefined;
        this.versionInfo = undefined;
    }
    render() {
        const { strings: { actions }, versionInfo } = this;
        const fullVersionName = versionInfo.versionIdentifier.name;
        const guid = versionInfo.versionIdentifier.guid;
        const actionElementId = `actionSwitchToVersion_${guid}`;
        return (h(Host, null, h("calcite-list-item", { description: fullVersionName.split(".")[0], label: fullVersionName.split(".")[1], value: guid }, h("calcite-action", { id: actionElementId, icon: "switch", text: "", onClick: () => {
                this.versionItemActionClickedEvent.emit({ actionType: "changeVersion", versionInfo });
            }, slot: "actions-start" }), h("calcite-tooltip", { "overlay-positioning": "fixed", referenceElement: actionElementId }, h("span", null, actions.switchToVersion)), h("calcite-action-menu", { "overlay-positioning": "fixed", slot: "actions-end", label: "" }, h("calcite-action", { icon: "switch", onClick: () => {
                this.versionItemActionClickedEvent.emit({ actionType: "changeVersion", versionInfo });
            }, scale: "s", text: actions.switchToVersion, textEnabled: true }), h("calcite-action", { disabled: fullVersionName.toUpperCase() === "SDE.DEFAULT" ||
                !this.hasAdvEditingUte ||
                !this.isVersioningApiAvailable, icon: "edit-attributes", onClick: () => {
                this.versionItemActionClickedEvent.emit({ actionType: "editVersion", versionInfo });
            }, scale: "s", text: actions.editVersion, textEnabled: true })))));
    }
    get _hostElement() { return getElement(this); }
};

const ArcgisVersionManagementVersionList = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.versionItemActionClickedEventWithServiceUrl = createEvent(this, "versionItemActionClickedEventWithServiceUrl", 7);
        this.versionListActionClickedEvent = createEvent(this, "versionListActionClickedEvent", 7);
        this._watchHandles = [];
        this.heading = undefined;
        this.serviceName = undefined;
        this.serviceUrl = undefined;
        this.strings = undefined;
        this.userHasAdvancedEditingUserTypeExtension = undefined;
        this.versionInfos = undefined;
        this.viewModel = undefined;
        this.state = undefined;
    }
    async componentDidLoad() {
        const { watch } = await importCoreReactiveUtils();
        this._watchHandles.push(watch(() => this.versionInfos, (versionInfos) => {
            this.versionInfos = versionInfos;
        }), watch(() => this.viewModel.state, (state) => {
            this.state = state;
        }));
    }
    async componentWillLoad() {
        this.state = this.viewModel.state;
    }
    disconnectedCallback() {
        this._watchHandles.forEach((handle) => {
            handle.remove();
        });
        this._watchHandles = [];
    }
    render() {
        const { heading, serviceName, serviceUrl, state, strings, versionInfos, viewModel } = this;
        // @ts-ignore
        const currentVersionIdentifier = viewModel.versionIdentifierLookup.get(serviceUrl);
        return (h(Host, null, h("calcite-flow-item", { heading: heading, description: serviceName }, currentVersionIdentifier ? (h("calcite-panel", { disabled: state === "executing", class: "panel-versions", heading: "Current version", description: currentVersionIdentifier.name })) : undefined, h("calcite-panel", { loading: state === "executing", class: "panel-versions", heading: versionInfos.length
                ? `${strings.headers.manageVersions} (${versionInfos.length})`
                : strings.headers.manageVersions }, h("calcite-action", { id: "actionRefreshVersions", icon: "refresh", text: "", onClick: () => {
                this.versionListActionClickedEvent.emit({
                    actionType: "refreshVersions",
                    serviceUrl
                });
            }, slot: "header-actions-end" }), h("calcite-tooltip", { "overlay-positioning": "fixed", referenceElement: "actionRefreshVersions" }, h("span", null, strings.actions.refreshVersions)), h("calcite-list", { class: "version-list", filterEnabled: true }, versionInfos.map((versionInfo) => (h("arcgis-version-management-version-item", { strings: strings, hasAdvEditingUte: viewModel.advancedEditingUserTypeExtensionLookup.get(serviceUrl), isVersioningApiAvailable: (viewModel.serverVersionLookup.get(serviceUrl) ?? 0) >= 11.2, versionInfo: versionInfo, onVersionItemActionClickedEvent: (e) => {
                this.versionItemActionClickedEventWithServiceUrl.emit({ ...e.detail, serviceUrl });
            } })))), h("calcite-fab", { appearance: "outline-fill", slot: "fab", textEnabled: true, text: strings.actions.newVersion, onClick: () => {
                this.versionListActionClickedEvent.emit({
                    actionType: "newVersion",
                    serviceUrl
                });
            } })))));
    }
    get _hostElement() { return getElement(this); }
};

export { ArcgisVersionManagementVersionItem as arcgis_version_management_version_item, ArcgisVersionManagementVersionList as arcgis_version_management_version_list };
