// @arcgis/map-components/dist/cjs/arcgis-placement.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-6e62126d.js');
const componentUtils = require('./component-utils-e47db275.js');
const disconnected = require('./disconnected-bc6a8942.js');

const placementCss = "arcgis-placement{}";

const ArcgisPlacement = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.placementReady = index.createEvent(this, "placementReady", 7);
        this.referenceElement = undefined;
        this.view = undefined;
        this.position = "bottom-left";
    }
    referenceElementWatcher(newValue) {
        void componentUtils.referenceElementHelper(this, newValue);
    }
    positionWatcher(newValue) {
        if (newValue !== undefined) {
            this.view.ui.move(this.childElem, newValue);
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        this.placementReady.emit({ ready: true });
        componentUtils.checkForView(this);
    }
    disconnectedCallback() {
        disconnected.onDisconnectedHelper(this);
    }
    render() {
        return (index.h("div", { ref: (el) => (this.childElem = el) }, index.h("slot", null)));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "position": ["positionWatcher"]
    }; }
};
ArcgisPlacement.style = placementCss;

exports.arcgis_placement = ArcgisPlacement;
