// @arcgis/map-components/dist/cjs/arcgis-utility-network-trace.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-6e62126d.js');
const index$1 = require('./index-b2323407.js');
const componentUtils = require('./component-utils-e47db275.js');
const disconnected = require('./disconnected-bc6a8942.js');

const utilityNetworkTraceCss = "";

const ArcgisUtilityNetworkTrace = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisUtilityNetworkTraceReady = index.createEvent(this, "arcgisUtilityNetworkTraceReady", 7);
        this.utilityNetworkTraceAddFlag = index.createEvent(this, "utilityNetworkTraceAddFlag", 7);
        this.utilityNetworkTraceAddFlagComplete = index.createEvent(this, "utilityNetworkTraceAddFlagComplete", 7);
        this.utilityNetworkTraceAddFlagError = index.createEvent(this, "utilityNetworkTraceAddFlagError", 7);
        this.utilityNetworkTraceAddResultArea = index.createEvent(this, "utilityNetworkTraceAddResultArea", 7);
        this.utilityNetworkTraceRemoveResultArea = index.createEvent(this, "utilityNetworkTraceRemoveResultArea", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.enableResultArea = false;
        this.defaultGraphicColor = {
            color: [255, 255, 0, 0.6],
            haloOpacity: 0.9,
            fillOpacity: 0.2,
            hex: "#FFFF00"
        };
        this.disabled = false;
        this.flags = [];
        this.gdbVersion = "sde.DEFAULT";
        this.inputSettings = [];
        this.resultAreaProperties = undefined;
        this.selectedTraces = [];
        this.selectOnComplete = true;
        this.showGraphicsOnComplete = true;
        this.showSelectionAttributes = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void componentUtils.referenceElementHelper(this, value);
    }
    enableResultAreaWatcher(value) {
        if (this.widget) {
            this.widget.enableResultArea = value;
        }
    }
    defaultGraphicColorWatcher(value) {
        if (this.widget) {
            this.widget.defaultGraphicColor = value;
        }
    }
    disabledWatcher(value, old) {
        if (this.widget && value !== old) {
            this.widget.disabled = value;
        }
    }
    flagsWatcher(value) {
        if (this.widget) {
            this.widget.flags = value;
        }
    }
    gdbVersionWatcher(value) {
        if (this.widget) {
            this.widget.gdbVersion = value;
        }
    }
    inputSettingsWatcher(value) {
        if (this.widget) {
            this.widget.inputSettings = value;
        }
    }
    resultAreaPropertiesWatcher(value) {
        if (this.widget) {
            this.widget.resultAreaProperties = value;
        }
    }
    selectedTracesWatcher(value) {
        if (this.widget) {
            this.widget.selectedTraces = value;
        }
    }
    selectOnCompleteWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.selectOnComplete = value;
        }
    }
    showGraphicsOnCompleteWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.showGraphicsOnComplete = value;
        }
    }
    showSelectionAttributesWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.showSelectionAttributes = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async checkCanTrace() {
        await this.widget?.checkCanTrace();
    }
    async confirmReset() {
        this.widget?.confirmReset();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await index$1.newWidgetsUtilityNetworkTrace(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisUtilityNetworkTraceReady.emit();
        componentUtils.checkForView(this);
    }
    disconnectedCallback() {
        disconnected.onDisconnectedHelper(this);
    }
    render() {
        return index.h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await index$1.importCoreReactiveUtils();
        componentUtils.addHandles(this, watch(() => this.widget.disabled, (value) => (this.disabled = value)));
    }
    _createParams() {
        const params = {};
        if (this.defaultGraphicColor) {
            params.defaultGraphicColor = this.defaultGraphicColor;
        }
        if (this.enableResultArea !== undefined) {
            params.enableResultArea = this.enableResultArea;
        }
        if (this.flags) {
            params.flags = this.flags;
        }
        if (this.gdbVersion) {
            params.gdbVersion = this.gdbVersion;
        }
        if (this.inputSettings) {
            params.inputSettings = this.inputSettings;
        }
        if (this.resultAreaProperties) {
            params.resultAreaProperties = this.resultAreaProperties;
        }
        if (this.selectedTraces) {
            params.selectedTraces = this.selectedTraces;
        }
        if (this.selectOnComplete !== undefined) {
            params.selectOnComplete = this.selectOnComplete;
        }
        if (this.showGraphicsOnComplete !== undefined) {
            params.showGraphicsOnComplete = this.showGraphicsOnComplete;
        }
        if (this.showSelectionAttributes !== undefined) {
            params.showSelectionAttributes = this.showSelectionAttributes;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("add-flag", (e) => this.utilityNetworkTraceAddFlag.emit(e));
        this.widget.on("add-flag-complete", (e) => this.utilityNetworkTraceAddFlagComplete.emit(e));
        this.widget.on("add-flag-error", (e) => this.utilityNetworkTraceAddFlagError.emit(e));
        this.widget.on("add-result-area", (e) => this.utilityNetworkTraceAddResultArea.emit(e));
        this.widget.on("remove-result-area", (e) => this.utilityNetworkTraceRemoveResultArea.emit(e));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "enableResultArea": ["enableResultAreaWatcher"],
        "defaultGraphicColor": ["defaultGraphicColorWatcher"],
        "disabled": ["disabledWatcher"],
        "flags": ["flagsWatcher"],
        "gdbVersion": ["gdbVersionWatcher"],
        "inputSettings": ["inputSettingsWatcher"],
        "resultAreaProperties": ["resultAreaPropertiesWatcher"],
        "selectedTraces": ["selectedTracesWatcher"],
        "selectOnComplete": ["selectOnCompleteWatcher"],
        "showGraphicsOnComplete": ["showGraphicsOnCompleteWatcher"],
        "showSelectionAttributes": ["showSelectionAttributesWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
};
ArcgisUtilityNetworkTrace.style = utilityNetworkTraceCss;

exports.arcgis_utility_network_trace = ArcgisUtilityNetworkTrace;
