// @arcgis/map-components/dist/cjs/arcgis-features.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-6e62126d.js');
const index$1 = require('./index-b2323407.js');
const componentUtils = require('./component-utils-e47db275.js');
const disconnected = require('./disconnected-bc6a8942.js');

const featuresCss = "arcgis-features{}";

const ArcgisFeatures = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisFeaturesReady = index.createEvent(this, "arcgisFeaturesReady", 7);
        this.arcgisPropertyChange = index.createEvent(this, "arcgisPropertyChange", 7);
        this.featuresTriggerAction = index.createEvent(this, "featuresTriggerAction", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.collapsed = false;
        this.content = undefined;
        this.features = undefined;
        this.icon = undefined;
        this.promises = undefined;
        this.selectedFeature = undefined;
        this.selectedFeatureIndex = undefined;
        this.featuresTitle = undefined;
        this.visibleElementsActionBar = true;
        this.visibleElementsCloseButton = true;
        this.visibleElementsCollapseButton = true;
        this.visibleElementsFeatureNavigation = true;
        this.visibleElementsHeading = true;
        this.visibleElementsSpinner = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void componentUtils.referenceElementHelper(this, value);
    }
    contentWatcher(value) {
        if (this.widget) {
            this.widget.content = value;
        }
    }
    featuresWatcher(value) {
        if (this.widget) {
            this.widget.features = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    featuresTitleWatcher(value) {
        if (this.widget) {
            this.widget.title = value;
        }
    }
    visibleElementsActionBarWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                actionBar: value
            };
        }
    }
    visibleElementsCloseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                closeButton: value
            };
        }
    }
    visibleElementsCollapseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                collapseButton: value
            };
        }
    }
    visibleElementsFeatureNavigationWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                featureNavigation: value
            };
        }
    }
    visibleElementsHeadingWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                heading: value
            };
        }
    }
    visibleElementsSpinnerWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                spinner: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async fetchFeatures(screenPoint, options) {
        return await this.widget?.fetchFeatures(screenPoint, options);
    }
    async clear() {
        this.widget?.clear();
    }
    async close() {
        this.widget?.close();
    }
    async next() {
        return this.widget?.next();
    }
    async open(options) {
        this.widget?.open(options);
    }
    async previous() {
        return this.widget?.previous();
    }
    async triggerAction(actionIndex) {
        this.widget?.triggerAction(actionIndex);
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await index$1.newWidgetsFeatures(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisFeaturesReady.emit();
        componentUtils.checkForView(this);
    }
    disconnectedCallback() {
        disconnected.onDisconnectedHelper(this);
    }
    render() {
        return index.h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await index$1.importCoreReactiveUtils();
        componentUtils.addHandles(this, watch(() => this.widget.collapsed, (value) => (this.collapsed = value), {
            initial: true
        }));
        componentUtils.addHandles(this, watch(() => this.widget.promises, (value) => (this.promises = value), {
            initial: true
        }));
        componentUtils.addHandles(this, watch(() => this.widget.selectedFeature, (value) => {
            this.selectedFeature = value;
            this.arcgisPropertyChange.emit({ name: "selectedFeature" });
        }, {
            initial: true
        }));
        componentUtils.addHandles(this, watch(() => this.widget.selectedFeatureIndex, (value) => {
            this.selectedFeatureIndex = value;
            this.arcgisPropertyChange.emit({ name: "selectedFeatureIndex" });
        }, {
            initial: true
        }));
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            actionBar: this.visibleElementsActionBar,
            closeButton: this.visibleElementsCloseButton,
            collapseButton: this.visibleElementsCollapseButton,
            featureNavigation: this.visibleElementsFeatureNavigation,
            heading: this.visibleElementsHeading,
            spinner: this.visibleElementsSpinner
        };
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.view) {
            params.view = this.view;
        }
        if (this.content) {
            params.content = this.content;
        }
        if (this.features) {
            params.features = this.features;
        }
        if (this.featuresTitle) {
            params.title = this.featuresTitle;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("trigger-action", (e) => this.featuresTriggerAction.emit(e));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "content": ["contentWatcher"],
        "features": ["featuresWatcher"],
        "icon": ["iconWatcher"],
        "featuresTitle": ["featuresTitleWatcher"],
        "visibleElementsActionBar": ["visibleElementsActionBarWatcher"],
        "visibleElementsCloseButton": ["visibleElementsCloseButtonWatcher"],
        "visibleElementsCollapseButton": ["visibleElementsCollapseButtonWatcher"],
        "visibleElementsFeatureNavigation": ["visibleElementsFeatureNavigationWatcher"],
        "visibleElementsHeading": ["visibleElementsHeadingWatcher"],
        "visibleElementsSpinner": ["visibleElementsSpinnerWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
};
ArcgisFeatures.style = featuresCss;

exports.arcgis_features = ArcgisFeatures;
