// @arcgis/map-components/dist/cjs/arcgis-version-management.cjs.entry.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-6e62126d.js');
const index$1 = require('./index-b2323407.js');

function S(e,t){let n=e;for(;n;){if(n===t)return !0;if(!n.parentNode)return !1;n.parentNode instanceof ShadowRoot?n=n.parentNode.host:n=n.parentNode;}return !1}function x(e,t,n=()=>{}){if(!t||t.length<=0)return;let r=new MutationObserver(o=>{for(let s of o)S(e,s.target)&&n();});return r.observe(document.documentElement,{attributes:!0,attributeFilter:t,subtree:!0}),r}function T(e,t){let n=e;for(;n;){let r=n.closest(t);if(r)return r;let o=n.getRootNode();if(o===document)return null;n=o.host;}return null}function f(e,t,n){return T(e,`[${t}]`)?.getAttribute(t)??n}var E=new Set(["ar","bg","bs","ca","cs","da","de","el","en","es","et","fi","fr","he","hr","hu","id","it","ja","ko","lt","lv","nl","nb","pl","pt-BR","pt-PT","ro","ru","sk","sl","sr","sv","th","tr","uk","vi","zh-CN","zh-HK","zh-TW"]),a=new Map;async function v(e,t,n=""){let r=`${t}/${n}${e}.json`;if(a.has(r))return a.get(r);try{let o=await fetch(r);if(o.ok){let i=await o.json();return a.set(r,i),i}if(e==="en"){a.set(r,void 0);return}let s=await v("en",t,n);return a.set(r,s),s}catch{a.set(r,void 0);return}}function _(e){let t=f(e,"lang",navigator.language||"en");if(E.has(t))return {lang:t,t9nLocale:t};let n=t.slice(0,2);return {lang:t,t9nLocale:E.has(n)?n:"en"}}var p=new WeakMap,m=new WeakMap,C=new WeakMap;async function B(e,t,n){m.set(e,t),p.set(e,x(e._hostElement,["lang"],()=>{b(e,n).catch(console.error);})),await b(e,n);}function q(e){p.get(e)?.disconnect(),p.delete(e),m.delete(e);}async function b(e,t){let{lang:n,t9nLocale:r}=_(e._hostElement);if(n===e._lang&&r===e._t9nLocale||(n!==e._lang&&(e._lang=n),r===e._t9nLocale))return;let o=C.get(e);if(!o){let s=m.get(e)??"",i=e._hostElement.tagName.toLowerCase().replace("arcgis-",""),l=`${s}/${i}/t9n`,u=`${i}.t9n.`;o=await v(r,l,u);}r!==e._t9nLocale&&(e._t9nLocale=r,e._t9nStrings=o,await t?.call(e));}

const versionManagementCss = ".div-content{background-color:var(--calcite-ui-foreground-1);padding:0.75rem}.flow-main{width:330px}.notice{width:-webkit-fill-available}.panel-services{max-height:400px}.panel-versions{max-height:400px}";

const ArcgisVersionManagement = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this._watchHandles = [];
        this.label = undefined;
        this.position = "top-right";
        this.view = undefined;
        this._lang = "";
        this._t9nLocale = "";
        this._t9nStrings = undefined;
        this.state = undefined;
    }
    positionWatcher(newValue) {
        if (newValue !== undefined && this.viewModel) {
            this.view.ui.move(this._hostElement, newValue);
        }
    }
    async componentDidLoad() {
        const { watch } = await index$1.importCoreReactiveUtils();
        this._watchHandles.push(watch(() => this.viewModel.state, (state) => {
            this.state = state;
        }));
    }
    async componentWillLoad() {
        await B(this, index.getAssetPath("./assets"));
        await this._checkForView();
        this.viewModel = await index$1.newWidgetsVersionManagementVersionManagementViewModel({
            view: this.view
        });
        this.label = this.label ? this.label : this._t9nStrings.label;
    }
    disconnectedCallback() {
        q(this);
        this._watchHandles.forEach((handle) => {
            handle.remove();
        });
        this._watchHandles = [];
    }
    render() {
        const { _t9nStrings, label, viewModel, viewModel: { state, loadError } } = this;
        const arrayServiceNames = Array.from(this.viewModel.serviceNameLookup, ([url, name]) => ({ url, name }));
        return (index.h(index.Host, null, index.h("calcite-flow", { ref: (node) => {
                this.flowElement = node;
            }, class: "flow-main" }, index.h("calcite-flow-item", { heading: label }, index.h("calcite-panel", { class: "panel-services", loading: state === "loading" }, arrayServiceNames.map((service) => (index.h("arcgis-version-management-service-item", { flowElement: this.flowElement, heading: label, serviceUrl: service.url, strings: _t9nStrings, viewModel: viewModel, onVersionItemActionClickedEventWithServiceUrl: (e) => {
                if (this.flowElement) {
                    this._handleVersionItemActionClick(e, this.flowElement);
                }
            }, onVersionListActionClickedEvent: (e) => {
                if (this.flowElement) {
                    this._handleVersionListActionClick(e, this.flowElement);
                }
            } }))), state === "disabled" ? (index.h("calcite-notice", { class: "notice", closable: false, kind: "warning", open: true, scale: "s", slot: "footer" }, index.h("div", { slot: "message" }, this._getLoadError(loadError)))) : undefined)))));
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    async _checkForView() {
        const viewRef = this._hostElement.closest("arcgis-map");
        if (viewRef?.view) {
            await viewRef.view.map.load();
            this.view = viewRef.view;
            this.view.ui.add(this._hostElement, this.position);
        }
    }
    _createVersionPropertiesFlow(serviceUrl, versionInfo) {
        const { flowElement, viewModel } = this;
        if (!flowElement) {
            return null;
        }
        const versionPropertiesFlow = document.createElement("arcgis-version-management-version-properties");
        versionPropertiesFlow.flowElement = flowElement;
        versionPropertiesFlow.serviceUrl = serviceUrl;
        versionPropertiesFlow.strings = this._t9nStrings;
        versionPropertiesFlow.versionInfo = versionInfo;
        versionPropertiesFlow.viewModel = viewModel;
        versionPropertiesFlow.addEventListener("versionPropertiesSaveClickedEvent", async (e) => {
            const { serviceUrl, versionInfo } = e.detail;
            const fullVersionName = versionInfo.versionIdentifier.name;
            const versionName = fullVersionName.split(".").length > 1 ? fullVersionName.split(".")[1] : fullVersionName;
            const ownerName = fullVersionName.split(".").length > 1 ? fullVersionName.split(".")[0] : viewModel.userLookup.get(serviceUrl);
            if (versionInfo.versionIdentifier.guid === "") {
                await viewModel.createVersion({
                    access: versionInfo.access,
                    description: versionInfo.description,
                    featureServerUrl: serviceUrl,
                    ownerName,
                    versionName
                });
            }
            else {
                await viewModel.alterVersion({
                    access: versionInfo.access,
                    description: versionInfo.description,
                    featureServerUrl: serviceUrl,
                    ownerName,
                    versionIdentifier: {
                        name: "",
                        guid: versionInfo.versionIdentifier.guid
                    },
                    versionName
                });
            }
            if (viewModel.state === "ready") {
                const versionList = flowElement.getElementsByTagName("arcgis-version-management-version-list")[0];
                versionList.versionInfos = [];
                const versionInfos = await viewModel.getVersionInfos(serviceUrl, true);
                versionList.versionInfos = versionInfos;
                await flowElement.back();
            }
        });
        return versionPropertiesFlow;
    }
    _getLoadError(loadError) {
        const { _t9nStrings } = this;
        switch (loadError) {
            case "no-feature-services":
                return _t9nStrings.loadErrors.noFeatureServices;
            case "no-layers-property":
                return _t9nStrings.loadErrors.noLayersProperty;
            default:
                return loadError;
        }
    }
    async _handleVersionItemActionClick(e, flowElement) {
        const { actionType, serviceUrl, versionInfo } = e.detail;
        switch (actionType) {
            case "changeVersion":
                await this._switchToVersion(serviceUrl, versionInfo.versionIdentifier.name, versionInfo.versionIdentifier.guid);
                break;
            case "editVersion": {
                const versionPropertiesFlow = this._createVersionPropertiesFlow(serviceUrl, versionInfo);
                if (versionPropertiesFlow) {
                    flowElement.append(versionPropertiesFlow);
                }
                break;
            }
        }
    }
    async _handleVersionListActionClick(e, flowElement) {
        const { actionType, serviceUrl } = e.detail;
        switch (actionType) {
            case "newVersion": {
                const versionPropertiesFlow = this._createVersionPropertiesFlow(serviceUrl, undefined);
                if (versionPropertiesFlow) {
                    flowElement.append(versionPropertiesFlow);
                }
                break;
            }
            case "refreshVersions": {
                if (this.flowElement) {
                    const versionList = this.flowElement.getElementsByTagName("arcgis-version-management-version-list")[0];
                    versionList.versionInfos = [];
                    const versionInfos = await this.viewModel.getVersionInfos(serviceUrl, true);
                    versionList.versionInfos = versionInfos;
                }
                break;
            }
        }
    }
    async _switchToVersion(featureServerUrl, toVersionName, toVersionGuid) {
        await this.viewModel.changeVersion(featureServerUrl, toVersionName, toVersionGuid);
    }
    static get assetsDirs() { return ["assets"]; }
    get _hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "position": ["positionWatcher"]
    }; }
};
ArcgisVersionManagement.style = versionManagementCss;

exports.arcgis_version_management = ArcgisVersionManagement;
