// @arcgis/map-components/dist/components/arcgis-elevation-profile.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { t as newWidgetsElevationProfile, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const elevationProfileCss = "";

const ArcgisElevationProfile$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisElevationProfile extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisElevationProfileReady = createEvent(this, "arcgisElevationProfileReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.effectiveUnits = undefined;
        this.highlightEnabled = true;
        this.hoveredChartPosition = undefined;
        this.icon = undefined;
        this.input = undefined;
        this.profiles = undefined;
        this.progress = undefined;
        this.state = "disabled";
        this.unit = undefined;
        this.uniformChartScaling = undefined;
        this.unitOptions = undefined;
        this.visibleElementsLegend = true;
        this.visibleElementsChart = true;
        this.visibleElementsClearButton = true;
        this.visibleElementsSettingsButton = true;
        this.visibleElementsSketchButton = true;
        this.visibleElementsSelectButton = true;
        this.visibleElementsUnitSelector = true;
        this.visibleElementsUniformChartScalingToggle = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    highlightEnabledWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.highlightEnabled = value;
        }
    }
    iconWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.icon = value;
        }
    }
    inputWatcher(value) {
        if (value && this.widget) {
            this.widget.input = value;
        }
    }
    profilesWatcher(value) {
        if (value && this.widget) {
            this.widget.profiles = value;
        }
    }
    unitWatcher(value) {
        if (value && this.widget) {
            this.widget.unit = value;
        }
    }
    uniformChartScalingWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.uniformChartScaling = value;
        }
    }
    unitOptionsWatcher(value) {
        if (value && this.widget) {
            this.widget.unitOptions = value;
        }
    }
    visibleElementsLegendWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                legend: value
            };
        }
    }
    visibleElementsChartWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                chart: value
            };
        }
    }
    visibleElementsClearButtonWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                clearButton: value
            };
        }
    }
    visibleElementsSettingsButtonWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                settingsButton: value
            };
        }
    }
    visibleElementsSketchButtonWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                sketchButton: value
            };
        }
    }
    visibleElementsSelectButtonWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                selectButton: value
            };
        }
    }
    visibleElementsUnitSelectorWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                unitSelector: value
            };
        }
    }
    visibleElementsUniformChartScalingToggleWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                uniformChartScalingToggle: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async start() {
        this.widget.viewModel.start();
    }
    async stop() {
        this.widget.viewModel.stop();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsElevationProfile(params);
        this.widget = widget;
        this._setViewModelParams();
        await this._addWatchers();
        this._addContainer();
        this.arcgisElevationProfileReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch, whenOnce } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.input, () => {
                this.input = this.widget.input;
                this.arcgisPropertyChange.emit({ name: "input" });
            }, {
                initial: true
            }),
            watch(() => this.widget.viewModel.effectiveUnits, () => (this.effectiveUnits = this.widget.viewModel.effectiveUnits), {
                initial: true
            }),
            watch(() => this.widget.viewModel.hoveredChartPosition, () => (this.hoveredChartPosition = this.widget.viewModel.hoveredChartPosition), {
                initial: true
            }),
            watch(() => this.widget.viewModel.progress, () => (this.progress = this.widget.viewModel.progress), {
                initial: true
            }),
            watch(() => this.widget.viewModel.state, () => (this.state = this.widget.viewModel.state), {
                initial: true
            })
        ]);
        await whenOnce(() => this.widget.unitOptions);
        this.unitOptions = this.widget.unitOptions;
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            legend: this.visibleElementsLegend,
            chart: this.visibleElementsChart,
            clearButton: this.visibleElementsClearButton,
            settingsButton: this.visibleElementsSettingsButton,
            sketchButton: this.visibleElementsSketchButton,
            selectButton: this.visibleElementsSelectButton,
            unitSelector: this.visibleElementsUnitSelector,
            uniformChartScalingToggle: this.visibleElementsUniformChartScalingToggle
        };
        if (this.view) {
            params.view = this.view;
        }
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.input) {
            params.input = this.input;
        }
        if (this.profiles) {
            params.profiles = this.profiles;
        }
        if (this.unit) {
            params.unit = this.unit;
        }
        if (this.unitOptions) {
            params.unitOptions = this.unitOptions;
        }
        return params;
    }
    _setViewModelParams() {
        const params = {};
        if (this.highlightEnabled !== undefined) {
            params.highlightEnabled = this.highlightEnabled;
        }
        if (this.uniformChartScaling !== undefined) {
            params.uniformChartScaling = this.uniformChartScaling;
        }
        this.widget.viewModel.set(params);
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "highlightEnabled": ["highlightEnabledWatcher"],
        "icon": ["iconWatcher"],
        "input": ["inputWatcher"],
        "profiles": ["profilesWatcher"],
        "unit": ["unitWatcher"],
        "uniformChartScaling": ["uniformChartScalingWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "visibleElementsLegend": ["visibleElementsLegendWatcher"],
        "visibleElementsChart": ["visibleElementsChartWatcher"],
        "visibleElementsClearButton": ["visibleElementsClearButtonWatcher"],
        "visibleElementsSettingsButton": ["visibleElementsSettingsButtonWatcher"],
        "visibleElementsSketchButton": ["visibleElementsSketchButtonWatcher"],
        "visibleElementsSelectButton": ["visibleElementsSelectButtonWatcher"],
        "visibleElementsUnitSelector": ["visibleElementsUnitSelectorWatcher"],
        "visibleElementsUniformChartScalingToggle": ["visibleElementsUniformChartScalingToggleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return elevationProfileCss; }
}, [0, "arcgis-elevation-profile", {
        "referenceElement": [1, "reference-element"],
        "effectiveUnits": [1040],
        "highlightEnabled": [4, "highlight-enabled"],
        "hoveredChartPosition": [1026, "hovered-chart-position"],
        "icon": [1],
        "input": [1040],
        "profiles": [16],
        "progress": [1026],
        "state": [1537],
        "unit": [1],
        "uniformChartScaling": [4, "uniform-chart-scaling"],
        "unitOptions": [16],
        "visibleElementsLegend": [4, "visible-elements-legend"],
        "visibleElementsChart": [4, "visible-elements-chart"],
        "visibleElementsClearButton": [4, "visible-elements-clear-button"],
        "visibleElementsSettingsButton": [4, "visible-elements-settings-button"],
        "visibleElementsSketchButton": [4, "visible-elements-sketch-button"],
        "visibleElementsSelectButton": [4, "visible-elements-select-button"],
        "visibleElementsUnitSelector": [4, "visible-elements-unit-selector"],
        "visibleElementsUniformChartScalingToggle": [4, "visible-elements-uniform-chart-scaling-toggle"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "start": [64],
        "stop": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "highlightEnabled": ["highlightEnabledWatcher"],
        "icon": ["iconWatcher"],
        "input": ["inputWatcher"],
        "profiles": ["profilesWatcher"],
        "unit": ["unitWatcher"],
        "uniformChartScaling": ["uniformChartScalingWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "visibleElementsLegend": ["visibleElementsLegendWatcher"],
        "visibleElementsChart": ["visibleElementsChartWatcher"],
        "visibleElementsClearButton": ["visibleElementsClearButtonWatcher"],
        "visibleElementsSettingsButton": ["visibleElementsSettingsButtonWatcher"],
        "visibleElementsSketchButton": ["visibleElementsSketchButtonWatcher"],
        "visibleElementsSelectButton": ["visibleElementsSelectButtonWatcher"],
        "visibleElementsUnitSelector": ["visibleElementsUnitSelectorWatcher"],
        "visibleElementsUniformChartScalingToggle": ["visibleElementsUniformChartScalingToggleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-elevation-profile"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-elevation-profile":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisElevationProfile$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisElevationProfile = ArcgisElevationProfile$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisElevationProfile, defineCustomElement };
