// @arcgis/map-components/dist/components/arcgis-track.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { $ as newWidgetsTrack, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const trackCss = "";

const ArcgisTrack$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisTrack extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisTrackReady = createEvent(this, "arcgisTrackReady", 7);
        this.trackComplete = createEvent(this, "trackComplete", 7);
        this.trackError = createEvent(this, "trackError", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.error = undefined;
        this.geolocationOptions = undefined;
        this.goToLocationEnabled = true;
        this.graphic = undefined;
        this.icon = undefined;
        this.positionFilterFunction = undefined;
        this.rotationEnabled = true;
        this.scale = undefined;
        this.state = "disabled";
        this.tracking = false;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    geolocationOptionsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.geolocationOptions = value;
        }
    }
    goToLocationEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.goToLocationEnabled = value;
        }
    }
    graphicWatcher(value) {
        if (this.widget) {
            this.widget.graphic = value;
        }
    }
    iconWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.icon = value;
        }
    }
    positionFilterFunctionWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.positionFilterFunction = value;
        }
    }
    rotationEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.rotationEnabled = value;
        }
    }
    scaleWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.scale = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async start() {
        this.widget.start();
    }
    async stop() {
        this.widget.stop();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsTrack(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisTrackReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.viewModel.error, (value) => (this.error = value)),
            watch(() => this.widget.viewModel.state, (value) => (this.state = value)),
            watch(() => this.widget.tracking, (value) => (this.tracking = value), {
                initial: true
            })
        ]);
    }
    _createParams() {
        const params = {};
        if (this.geolocationOptions !== undefined) {
            params.geolocationOptions = this.geolocationOptions;
        }
        if (this.goToLocationEnabled !== undefined) {
            params.goToLocationEnabled = this.goToLocationEnabled;
        }
        if (this.graphic !== undefined) {
            params.graphic = this.graphic;
        }
        if (this.icon !== undefined) {
            params.icon = this.icon;
        }
        if (this.positionFilterFunction !== undefined) {
            params.viewModel = {
                positionFilterFunction: this.positionFilterFunction
            };
        }
        if (this.rotationEnabled !== undefined) {
            params.rotationEnabled = this.rotationEnabled;
        }
        if (this.scale !== undefined) {
            params.scale = this.scale;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("track", (e) => this.trackComplete.emit(e));
        this.widget.on("track-error", (e) => this.trackError.emit(e));
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "geolocationOptions": ["geolocationOptionsWatcher"],
        "goToLocationEnabled": ["goToLocationEnabledWatcher"],
        "graphic": ["graphicWatcher"],
        "icon": ["iconWatcher"],
        "positionFilterFunction": ["positionFilterFunctionWatcher"],
        "rotationEnabled": ["rotationEnabledWatcher"],
        "scale": ["scaleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return trackCss; }
}, [0, "arcgis-track", {
        "referenceElement": [1, "reference-element"],
        "error": [1032],
        "geolocationOptions": [8, "geolocation-options"],
        "goToLocationEnabled": [4, "go-to-location-enabled"],
        "graphic": [16],
        "icon": [1],
        "positionFilterFunction": [16],
        "rotationEnabled": [4, "rotation-enabled"],
        "scale": [2],
        "state": [1537],
        "tracking": [1540],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "start": [64],
        "stop": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "geolocationOptions": ["geolocationOptionsWatcher"],
        "goToLocationEnabled": ["goToLocationEnabledWatcher"],
        "graphic": ["graphicWatcher"],
        "icon": ["iconWatcher"],
        "positionFilterFunction": ["positionFilterFunctionWatcher"],
        "rotationEnabled": ["rotationEnabledWatcher"],
        "scale": ["scaleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-track"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-track":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisTrack$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisTrack = ArcgisTrack$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisTrack, defineCustomElement };
