// @arcgis/map-components/dist/components/arcgis-directional-pad.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { m as newWidgetsDirectionalPad } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const directionalPadCss = "arcgis-directional-pad{}";

const ArcgisDirectionalPad$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisDirectionalPad extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisDirectionalPadReady = createEvent(this, "arcgisDirectionalPadReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.icon = undefined;
        this.visualScale = "s";
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    iconWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.icon = value;
        }
    }
    visualScaleWatcher(value) {
        if (this.widget) {
            this.widget.visualScale = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Watchers
    //
    //--------------------------------------------------------------------------
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsDirectionalPad(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisDirectionalPadReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.visualScale) {
            params.visualScale = this.visualScale;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "visualScale": ["visualScaleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return directionalPadCss; }
}, [0, "arcgis-directional-pad", {
        "referenceElement": [1, "reference-element"],
        "icon": [1],
        "visualScale": [1, "visual-scale"],
        "position": [1],
        "view": [16],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "visualScale": ["visualScaleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-directional-pad"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-directional-pad":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisDirectionalPad$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisDirectionalPad = ArcgisDirectionalPad$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisDirectionalPad, defineCustomElement };
