// @arcgis/map-components/dist/components/arcgis-swipe.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { Y as newWidgetsSwipe } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const swipeCss = "arcgis-swipe{}";

const ArcgisSwipe$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisSwipe extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.widgetReady = createEvent(this, "widgetReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.direction = undefined;
        this.dragLabel = undefined;
        this.leadingLayers = undefined;
        this.swipePosition = undefined;
        this.trailingLayers = undefined;
        this.view = undefined;
        this.position = "bottom-left";
        this.widget = undefined;
    }
    //--------------------------------------------------------------------------
    //
    //  Watchers
    //
    //--------------------------------------------------------------------------
    directionWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.direction = newValue;
        }
    }
    dragLabelWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.dragLabel = newValue;
        }
    }
    leadingLayersWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.leadingLayers = newValue;
        }
    }
    swipePositionWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.position = newValue;
        }
    }
    trailingLayersWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.trailingLayers = newValue;
        }
    }
    positionWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            if (typeof this.view !== "string") {
                this.view.ui.move(this.childElem, newValue);
            }
        }
    }
    referenceElementWatcher(newValue) {
        referenceElementHelper(this, newValue);
    }
    viewWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.view = newValue;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsSwipe(params);
        this.widget = widget;
        this.widgetReady.emit({ widget: this.widget });
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _createParams() {
        const params = {
            container: this.childElem
        };
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "direction": ["directionWatcher"],
        "dragLabel": ["dragLabelWatcher"],
        "leadingLayers": ["leadingLayersWatcher"],
        "swipePosition": ["swipePositionWatcher"],
        "trailingLayers": ["trailingLayersWatcher"],
        "position": ["positionWatcher"],
        "referenceElement": ["referenceElementWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return swipeCss; }
}, [0, "arcgis-swipe", {
        "referenceElement": [1, "reference-element"],
        "direction": [1],
        "dragLabel": [1, "drag-label"],
        "leadingLayers": [16],
        "swipePosition": [2, "swipe-position"],
        "trailingLayers": [16],
        "view": [1040],
        "position": [1],
        "widget": [1040]
    }, undefined, {
        "direction": ["directionWatcher"],
        "dragLabel": ["dragLabelWatcher"],
        "leadingLayers": ["leadingLayersWatcher"],
        "swipePosition": ["swipePositionWatcher"],
        "trailingLayers": ["trailingLayersWatcher"],
        "position": ["positionWatcher"],
        "referenceElement": ["referenceElementWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-swipe"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-swipe":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisSwipe$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisSwipe = ArcgisSwipe$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisSwipe, defineCustomElement };
