// @arcgis/map-components/dist/components/arcgis-directions.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { o as newLayersRouteLayer, p as newWidgetsDirections, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, i as isNullOrEmptyString, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const directionsCss = "";

const ArcgisDirections$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisDirections extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisDirectionsReady = createEvent(this, "arcgisDirectionsReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.apiKey = undefined;
        this.icon = undefined;
        this.routeLayerItemId = undefined;
        this.lastRoute = undefined;
        this.layer = undefined;
        this.maxStops = 50;
        this.searchProperties = undefined;
        this.unit = undefined;
        this.useDefaultRouteLayer = true;
        this.visibleElementsLayerDetails = true;
        this.visibleElementsSaveAsButton = true;
        this.visibleElementsSaveButton = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    apiKeyWatcher(value) {
        if (value && this.widget) {
            this.widget.apiKey = value;
        }
    }
    iconWatcher(value) {
        if (value && this.widget) {
            this.widget.icon = value;
        }
    }
    routeLayerItemIdWatcher(value) {
        if (isNullOrEmptyString(value)) {
            this.view?.map.remove(this.layer);
            // maintain a default route layer
            if (this.useDefaultRouteLayer) {
                void newLayersRouteLayer({}).then((layer) => {
                    this.layer = layer;
                });
            }
        }
        else if (value && this.widget) {
            void newLayersRouteLayer({
                portalItem: {
                    id: value
                }
            }).then((layer) => {
                this.layer = layer;
            });
        }
    }
    layerWatcher(value, old) {
        if (old) {
            this.view?.map.remove(old);
        }
        if (value && this.widget) {
            this.widget.layer = value;
            this.view?.map.add(this.layer);
        }
        else if (!value) {
            this.widget.layer = null;
        }
    }
    maxStopsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.maxStops = value;
        }
    }
    searchPropertiesWatcher(value) {
        if (this.widget) {
            this.widget.searchProperties = value;
        }
    }
    unitWatcher(value) {
        if (this.widget) {
            this.widget.unit = value;
        }
    }
    visibleElementsLayerDetailsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                layerDetails: value
            };
        }
    }
    visibleElementsSaveAsButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                saveAsButton: value
            };
        }
    }
    visibleElementsSaveButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                saveButton: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
            if (this.layer) {
                this.widget.view.map.add(this.layer);
            }
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async getDirections() {
        return await this.widget?.getDirections();
    }
    async save() {
        return await this.widget?.save();
    }
    async saveAs(portalItem, options) {
        return await this.widget?.saveAs(portalItem, options);
    }
    async zoomToRoute() {
        this.widget?.zoomToRoute();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = await this._createParams();
        const widget = await newWidgetsDirections(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisDirectionsReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.lastRoute, (value) => {
            this.lastRoute = value;
            this.arcgisPropertyChange.emit({ name: "lastRoute" });
        }, {
            initial: true
        }));
    }
    async _createParams() {
        const params = {};
        params.visibleElements = {
            layerDetails: this.visibleElementsLayerDetails,
            saveAsButton: this.visibleElementsSaveAsButton,
            saveButton: this.visibleElementsSaveButton
        };
        if (this.view) {
            params.view = this.view;
        }
        if (this.apiKey) {
            params.apiKey = this.apiKey;
        }
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.routeLayerItemId) {
            params.layer = await newLayersRouteLayer({
                portalItem: {
                    id: this.routeLayerItemId
                }
            });
            this.layer = params.layer;
        }
        else if (this.layer) {
            params.layer = this.layer;
            this.layer = params.layer;
        }
        else if (this.useDefaultRouteLayer) {
            params.layer = await newLayersRouteLayer({});
            this.layer = params.layer;
        }
        if (this.layer !== undefined) {
            this.view?.map.add(this.layer);
        }
        if (this.maxStops !== undefined) {
            params.maxStops = this.maxStops;
        }
        if (this.searchProperties) {
            params.searchProperties = this.searchProperties;
        }
        if (this.unit) {
            params.unit = this.unit;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "apiKey": ["apiKeyWatcher"],
        "icon": ["iconWatcher"],
        "routeLayerItemId": ["routeLayerItemIdWatcher"],
        "layer": ["layerWatcher"],
        "maxStops": ["maxStopsWatcher"],
        "searchProperties": ["searchPropertiesWatcher"],
        "unit": ["unitWatcher"],
        "visibleElementsLayerDetails": ["visibleElementsLayerDetailsWatcher"],
        "visibleElementsSaveAsButton": ["visibleElementsSaveAsButtonWatcher"],
        "visibleElementsSaveButton": ["visibleElementsSaveButtonWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return directionsCss; }
}, [0, "arcgis-directions", {
        "referenceElement": [1, "reference-element"],
        "apiKey": [1, "api-key"],
        "icon": [1],
        "routeLayerItemId": [1, "route-layer-item-id"],
        "lastRoute": [1040],
        "layer": [16],
        "maxStops": [2, "max-stops"],
        "searchProperties": [16],
        "unit": [1],
        "useDefaultRouteLayer": [4, "use-default-route-layer"],
        "visibleElementsLayerDetails": [4, "visible-elements-layer-details"],
        "visibleElementsSaveAsButton": [4, "visible-elements-save-as-button"],
        "visibleElementsSaveButton": [4, "visible-elements-save-button"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "getDirections": [64],
        "save": [64],
        "saveAs": [64],
        "zoomToRoute": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "apiKey": ["apiKeyWatcher"],
        "icon": ["iconWatcher"],
        "routeLayerItemId": ["routeLayerItemIdWatcher"],
        "layer": ["layerWatcher"],
        "maxStops": ["maxStopsWatcher"],
        "searchProperties": ["searchPropertiesWatcher"],
        "unit": ["unitWatcher"],
        "visibleElementsLayerDetails": ["visibleElementsLayerDetailsWatcher"],
        "visibleElementsSaveAsButton": ["visibleElementsSaveAsButtonWatcher"],
        "visibleElementsSaveButton": ["visibleElementsSaveButtonWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-directions"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-directions":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisDirections$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisDirections = ArcgisDirections$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisDirections, defineCustomElement };
