// @arcgis/map-components/dist/components/arcgis-search.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { S as newWidgetsSearch, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const searchCss = "";

const ArcgisSearch$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisSearch extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisSearchReady = createEvent(this, "arcgisSearchReady", 7);
        this.searchBlur = createEvent(this, "searchBlur", 7);
        this.searchClear = createEvent(this, "searchClear", 7);
        this.searchComplete = createEvent(this, "searchComplete", 7);
        this.searchFocus = createEvent(this, "searchFocus", 7);
        this.searchStart = createEvent(this, "searchStart", 7);
        this.searchSelectResult = createEvent(this, "searchSelectResult", 7);
        this.searchSuggestComplete = createEvent(this, "searchSuggestComplete", 7);
        this.searchSuggestStart = createEvent(this, "searchSuggestStart", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.activeMenu = "none";
        this.activeSource = undefined;
        this.activeSourceIndex = 0;
        this.allPlaceholder = "Find address or place";
        this.allSources = undefined;
        this.autoSelect = true;
        this.defaultSources = undefined;
        this.disabled = false;
        this.icon = undefined;
        this.includeDefaultSources = true;
        this.locationEnabled = true;
        this.maxResults = 6;
        this.maxSuggestions = 6;
        this.minSuggestCharacters = 3;
        this.popupEnabled = true;
        this.popupTemplate = undefined;
        this.portal = undefined;
        this.resultGraphic = undefined;
        this.resultGraphicEnabled = true;
        this.results = undefined;
        this.searchAllEnabled = true;
        this.searchTerm = undefined;
        this.selectedResult = undefined;
        this.sources = undefined;
        this.suggestions = undefined;
        this.suggestionsEnabled = true;
        this.position = "top-right";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    activeMenuWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.activeMenu = value;
        }
    }
    activeSourceIndexWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.activeSourceIndex = value;
        }
    }
    allPlaceholderWatcher(value) {
        if (this.widget) {
            this.widget.allPlaceholder = value;
        }
    }
    autoSelectWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.autoSelect = value;
        }
    }
    disabledWatcher(value, old) {
        if (this.widget && value !== old) {
            this.widget.disabled = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    includeDefaultSourcesWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.includeDefaultSources = value;
        }
    }
    locationEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.locationEnabled = value;
        }
    }
    maxResultsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.maxResults = value;
        }
    }
    maxSuggestionsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.maxSuggestions = value;
        }
    }
    minSuggestCharactersWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.minSuggestCharacters = value;
        }
    }
    popupEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.popupEnabled = value;
        }
    }
    popupTemplateWatcher(value) {
        if (this.widget) {
            this.widget.popupTemplate = value;
        }
    }
    portalWatcher(value) {
        if (this.widget) {
            this.widget.portal = value;
        }
    }
    resultGraphicEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.resultGraphicEnabled = value;
        }
    }
    searchAllEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.searchAllEnabled = value;
        }
    }
    searchTermWatcher(value) {
        if (this.widget) {
            this.widget.searchTerm = value;
        }
    }
    sourcesWatcher(value) {
        if (this.widget) {
            this.widget.sources = value;
        }
    }
    suggestionsEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.suggestionsEnabled = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async blurSearch() {
        this.widget?.blur();
    }
    async clearSearch() {
        this.widget?.clear();
    }
    async focusSearch() {
        this.widget?.focus();
    }
    async search(term) {
        return await this.widget?.search(term);
    }
    async suggest(term) {
        return await this.widget?.suggest(term);
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsSearch(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisSearchReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Interal Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.activeSource, (value) => (this.activeSource = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.activeSourceIndex, (value) => (this.activeSourceIndex = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.allSources, (value) => (this.allSources = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.defaultSources, (value) => (this.defaultSources = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.disabled, (value) => (this.disabled = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.resultGraphic, (value) => (this.resultGraphic = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.results, (value) => (this.results = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.selectedResult, (value) => (this.selectedResult = value), {
            initial: true
        }));
        addHandles(this, watch(() => this.widget.suggestions, (value) => (this.suggestions = value), {
            initial: true
        }));
    }
    _createParams() {
        const params = {};
        if (this.allPlaceholder) {
            params.allPlaceholder = this.allPlaceholder;
        }
        if (this.autoSelect !== undefined) {
            params.autoSelect = this.autoSelect;
        }
        if (this.includeDefaultSources !== undefined) {
            params.includeDefaultSources = this.includeDefaultSources;
        }
        if (this.locationEnabled !== undefined) {
            params.locationEnabled = this.locationEnabled;
        }
        if (this.maxResults !== undefined) {
            params.maxResults = this.maxResults;
        }
        if (this.maxSuggestions !== undefined) {
            params.maxSuggestions = this.maxSuggestions;
        }
        if (this.minSuggestCharacters !== undefined) {
            params.minSuggestCharacters = this.minSuggestCharacters;
        }
        if (this.popupEnabled !== undefined) {
            params.popupEnabled = this.popupEnabled;
        }
        if (this.popupTemplate) {
            params.popupTemplate = this.popupTemplate;
        }
        if (this.portal) {
            params.portal = this.portal;
        }
        if (this.resultGraphicEnabled !== undefined) {
            params.resultGraphicEnabled = this.resultGraphicEnabled;
        }
        if (this.searchAllEnabled !== undefined) {
            params.searchAllEnabled = this.searchAllEnabled;
        }
        if (this.searchTerm) {
            params.searchTerm = this.searchTerm;
        }
        if (this.sources) {
            params.sources = this.sources;
        }
        if (this.suggestionsEnabled !== undefined) {
            params.suggestionsEnabled = this.suggestionsEnabled;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("search-blur", (e) => this.searchBlur.emit(e));
        this.widget.on("search-clear", (e) => this.searchClear.emit(e));
        this.widget.on("search-complete", (e) => this.searchComplete.emit(e));
        this.widget.on("search-focus", (e) => this.searchFocus.emit(e));
        this.widget.on("search-start", (e) => this.searchStart.emit(e));
        this.widget.on("select-result", (e) => this.searchSelectResult.emit(e));
        this.widget.on("suggest-complete", (e) => this.searchSuggestComplete.emit(e));
        this.widget.on("suggest-start", (e) => this.searchSuggestStart.emit(e));
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "activeMenu": ["activeMenuWatcher"],
        "activeSourceIndex": ["activeSourceIndexWatcher"],
        "allPlaceholder": ["allPlaceholderWatcher"],
        "autoSelect": ["autoSelectWatcher"],
        "disabled": ["disabledWatcher"],
        "icon": ["iconWatcher"],
        "includeDefaultSources": ["includeDefaultSourcesWatcher"],
        "locationEnabled": ["locationEnabledWatcher"],
        "maxResults": ["maxResultsWatcher"],
        "maxSuggestions": ["maxSuggestionsWatcher"],
        "minSuggestCharacters": ["minSuggestCharactersWatcher"],
        "popupEnabled": ["popupEnabledWatcher"],
        "popupTemplate": ["popupTemplateWatcher"],
        "portal": ["portalWatcher"],
        "resultGraphicEnabled": ["resultGraphicEnabledWatcher"],
        "searchAllEnabled": ["searchAllEnabledWatcher"],
        "searchTerm": ["searchTermWatcher"],
        "sources": ["sourcesWatcher"],
        "suggestionsEnabled": ["suggestionsEnabledWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return searchCss; }
}, [0, "arcgis-search", {
        "referenceElement": [1, "reference-element"],
        "activeMenu": [1, "active-menu"],
        "activeSource": [1040],
        "activeSourceIndex": [1026, "active-source-index"],
        "allPlaceholder": [1, "all-placeholder"],
        "allSources": [1040],
        "autoSelect": [4, "auto-select"],
        "defaultSources": [1040],
        "disabled": [1540],
        "icon": [1],
        "includeDefaultSources": [4, "include-default-sources"],
        "locationEnabled": [4, "location-enabled"],
        "maxResults": [2, "max-results"],
        "maxSuggestions": [2, "max-suggestions"],
        "minSuggestCharacters": [2, "min-suggest-characters"],
        "popupEnabled": [4, "popup-enabled"],
        "popupTemplate": [16],
        "portal": [16],
        "resultGraphic": [1040],
        "resultGraphicEnabled": [4, "result-graphic-enabled"],
        "results": [1040],
        "searchAllEnabled": [4, "search-all-enabled"],
        "searchTerm": [1, "search-term"],
        "selectedResult": [1040],
        "sources": [16],
        "suggestions": [1040],
        "suggestionsEnabled": [4, "suggestions-enabled"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "blurSearch": [64],
        "clearSearch": [64],
        "focusSearch": [64],
        "search": [64],
        "suggest": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "activeMenu": ["activeMenuWatcher"],
        "activeSourceIndex": ["activeSourceIndexWatcher"],
        "allPlaceholder": ["allPlaceholderWatcher"],
        "autoSelect": ["autoSelectWatcher"],
        "disabled": ["disabledWatcher"],
        "icon": ["iconWatcher"],
        "includeDefaultSources": ["includeDefaultSourcesWatcher"],
        "locationEnabled": ["locationEnabledWatcher"],
        "maxResults": ["maxResultsWatcher"],
        "maxSuggestions": ["maxSuggestionsWatcher"],
        "minSuggestCharacters": ["minSuggestCharactersWatcher"],
        "popupEnabled": ["popupEnabledWatcher"],
        "popupTemplate": ["popupTemplateWatcher"],
        "portal": ["portalWatcher"],
        "resultGraphicEnabled": ["resultGraphicEnabledWatcher"],
        "searchAllEnabled": ["searchAllEnabledWatcher"],
        "searchTerm": ["searchTermWatcher"],
        "sources": ["sourcesWatcher"],
        "suggestionsEnabled": ["suggestionsEnabledWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-search"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-search":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisSearch$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisSearch = ArcgisSearch$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisSearch, defineCustomElement };
