// @arcgis/map-components/dist/components/arcgis-basemap-gallery.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { d as newWidgetsBasemapGallery, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const basemapGalleryCss = "";

const ArcgisBasemapGallery$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisBasemapGallery extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisBasemapGalleryReady = createEvent(this, "arcgisBasemapGalleryReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.activeBasemap = undefined;
        this.disabled = false;
        this.icon = undefined;
        this.source = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    disabledWatcher(value, old) {
        if (this.widget && value !== old) {
            this.widget.disabled = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    sourceWatcher(value) {
        if (this.widget) {
            this.widget.source = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsBasemapGallery(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisBasemapGalleryReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Interal Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.source, () => (this.source = this.widget.source), {
                initial: true
            }),
            watch(() => this.widget.activeBasemap, (basemap) => {
                this.activeBasemap = basemap;
                this.arcgisPropertyChange.emit({ name: "activeBasemap" });
            }, {
                initial: true
            }),
            watch(() => this.widget.disabled, (value) => (this.disabled = value))
        ]);
    }
    _createParams() {
        const params = {};
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.source) {
            params.source = this.source;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "disabled": ["disabledWatcher"],
        "icon": ["iconWatcher"],
        "source": ["sourceWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return basemapGalleryCss; }
}, [0, "arcgis-basemap-gallery", {
        "referenceElement": [1, "reference-element"],
        "activeBasemap": [1040],
        "disabled": [1540],
        "icon": [1],
        "source": [1040],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "disabled": ["disabledWatcher"],
        "icon": ["iconWatcher"],
        "source": ["sourceWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-basemap-gallery"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-basemap-gallery":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisBasemapGallery$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisBasemapGallery = ArcgisBasemapGallery$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisBasemapGallery, defineCustomElement };
