// @arcgis/map-components/dist/components/arcgis-bookmarks.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { g as newWidgetsBookmarks, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const bookmarksCss = "";

const ArcgisBookmarks$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisBookmarks extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisBookmarksReady = createEvent(this, "arcgisBookmarksReady", 7);
        this.bookmarkEdit = createEvent(this, "bookmarkEdit", 7);
        this.bookmarkSelect = createEvent(this, "bookmarkSelect", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.bookmarks = undefined;
        this.defaultCreateOptions = undefined;
        this.defaultEditOptions = undefined;
        this.disabled = false;
        this.dragEnabled = false;
        this.filterPlaceholder = undefined;
        this.filterText = undefined;
        this.icon = undefined;
        this.visibleElementsAddBookmarkButton = false;
        this.visibleElementsCloseButton = false;
        this.visibleElementsCollapseButton = false;
        this.visibleElementsEditBookmarkButton = false;
        this.visibleElementsFilter = false;
        this.visibleElementsFlow = true;
        this.visibleElementsHeading = false;
        this.visibleElementsThumbnail = true;
        this.visibleElementsTime = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    bookmarksWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.bookmarks = value;
        }
    }
    defaultCreateOptionsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.defaultCreateOptions = value;
        }
    }
    defaultEditOptionsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.defaultEditOptions = value;
        }
    }
    disabledWatcher(value, old) {
        if (this.widget && value !== old) {
            this.widget.disabled = value;
        }
    }
    dragEnabledWatcher(value) {
        if (this.widget) {
            this.widget.dragEnabled = value;
        }
    }
    filterPlaceholderWatcher(value) {
        if (this.widget) {
            this.widget.filterPlaceholder = value;
        }
    }
    filterTextWatcher(value) {
        if (this.widget) {
            this.widget.filterText = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    visibleElementsAddBookmarkButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, addBookmarkButton: value };
        }
    }
    visibleElementsCloseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, closeButton: value };
        }
    }
    visibleElementsCollapseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, collapseButton: value };
        }
    }
    visibleElementsEditBookmarkButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, editBookmarkButton: value };
        }
    }
    visibleElementsFilterWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, filter: value };
        }
    }
    visibleElementsFlowWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, flow: value };
        }
    }
    visibleElementsHeadingWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, heading: value };
        }
    }
    visibleElementsThumbnailWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, thumbnail: value };
        }
    }
    visibleElementsTimeWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = { ...this.widget.visibleElements, time: value };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsBookmarks(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisBookmarksReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.disabled, (value) => (this.disabled = value))
        ]);
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            addBookmarkButton: this.visibleElementsAddBookmarkButton,
            closeButton: this.visibleElementsCloseButton,
            collapseButton: this.visibleElementsCollapseButton,
            editBookmarkButton: this.visibleElementsEditBookmarkButton,
            filter: this.visibleElementsFilter,
            flow: this.visibleElementsFlow,
            heading: this.visibleElementsHeading,
            thumbnail: this.visibleElementsThumbnail,
            time: this.visibleElementsTime
        };
        if (this.view) {
            params.view = this.view;
        }
        if (this.bookmarks) {
            params.bookmarks = this.bookmarks;
        }
        if (this.defaultCreateOptions) {
            params.defaultCreateOptions = this.defaultCreateOptions;
        }
        if (this.defaultEditOptions) {
            params.defaultEditOptions = this.defaultEditOptions;
        }
        if (this.dragEnabled !== undefined) {
            params.dragEnabled = this.dragEnabled;
        }
        if (this.filterPlaceholder) {
            params.filterPlaceholder = this.filterPlaceholder;
        }
        if (this.filterText) {
            params.filterText = this.filterText;
        }
        if (this.icon) {
            params.icon = this.icon;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("bookmark-edit", (e) => this.bookmarkEdit.emit(e));
        this.widget.on("bookmark-select", (e) => this.bookmarkSelect.emit(e));
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "bookmarks": ["bookmarksWatcher"],
        "defaultCreateOptions": ["defaultCreateOptionsWatcher"],
        "defaultEditOptions": ["defaultEditOptionsWatcher"],
        "disabled": ["disabledWatcher"],
        "dragEnabled": ["dragEnabledWatcher"],
        "filterPlaceholder": ["filterPlaceholderWatcher"],
        "filterText": ["filterTextWatcher"],
        "icon": ["iconWatcher"],
        "visibleElementsAddBookmarkButton": ["visibleElementsAddBookmarkButtonWatcher"],
        "visibleElementsCloseButton": ["visibleElementsCloseButtonWatcher"],
        "visibleElementsCollapseButton": ["visibleElementsCollapseButtonWatcher"],
        "visibleElementsEditBookmarkButton": ["visibleElementsEditBookmarkButtonWatcher"],
        "visibleElementsFilter": ["visibleElementsFilterWatcher"],
        "visibleElementsFlow": ["visibleElementsFlowWatcher"],
        "visibleElementsHeading": ["visibleElementsHeadingWatcher"],
        "visibleElementsThumbnail": ["visibleElementsThumbnailWatcher"],
        "visibleElementsTime": ["visibleElementsTimeWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return bookmarksCss; }
}, [0, "arcgis-bookmarks", {
        "referenceElement": [1, "reference-element"],
        "bookmarks": [16],
        "defaultCreateOptions": [16],
        "defaultEditOptions": [16],
        "disabled": [1540],
        "dragEnabled": [4, "drag-enabled"],
        "filterPlaceholder": [1, "filter-placeholder"],
        "filterText": [1, "filter-text"],
        "icon": [1],
        "visibleElementsAddBookmarkButton": [4, "visible-elements-add-bookmark-button"],
        "visibleElementsCloseButton": [4, "visible-elements-close-button"],
        "visibleElementsCollapseButton": [4, "visible-elements-collapse-button"],
        "visibleElementsEditBookmarkButton": [4, "visible-elements-edit-bookmark-button"],
        "visibleElementsFilter": [4, "visible-elements-filter"],
        "visibleElementsFlow": [4, "visible-elements-flow"],
        "visibleElementsHeading": [4, "visible-elements-heading"],
        "visibleElementsThumbnail": [4, "visible-elements-thumbnail"],
        "visibleElementsTime": [4, "visible-elements-time"],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "bookmarks": ["bookmarksWatcher"],
        "defaultCreateOptions": ["defaultCreateOptionsWatcher"],
        "defaultEditOptions": ["defaultEditOptionsWatcher"],
        "disabled": ["disabledWatcher"],
        "dragEnabled": ["dragEnabledWatcher"],
        "filterPlaceholder": ["filterPlaceholderWatcher"],
        "filterText": ["filterTextWatcher"],
        "icon": ["iconWatcher"],
        "visibleElementsAddBookmarkButton": ["visibleElementsAddBookmarkButtonWatcher"],
        "visibleElementsCloseButton": ["visibleElementsCloseButtonWatcher"],
        "visibleElementsCollapseButton": ["visibleElementsCollapseButtonWatcher"],
        "visibleElementsEditBookmarkButton": ["visibleElementsEditBookmarkButtonWatcher"],
        "visibleElementsFilter": ["visibleElementsFilterWatcher"],
        "visibleElementsFlow": ["visibleElementsFlowWatcher"],
        "visibleElementsHeading": ["visibleElementsHeadingWatcher"],
        "visibleElementsThumbnail": ["visibleElementsThumbnailWatcher"],
        "visibleElementsTime": ["visibleElementsTimeWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-bookmarks"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-bookmarks":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisBookmarks$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisBookmarks = ArcgisBookmarks$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisBookmarks, defineCustomElement };
