// @arcgis/map-components/dist/components/arcgis-navigation-toggle.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { M as newWidgetsNavigationToggle } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const navigationToggleCss = "arcgis-navigation-toggle{}";

const ArcgisNavigationToggle$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisNavigationToggle extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisNavigationToggleReady = createEvent(this, "arcgisNavigationToggleReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.layout = "vertical";
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    layoutWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.layout = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Watchers
    //
    //--------------------------------------------------------------------------
    async toggle() {
        this.widget?.toggle();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsNavigationToggle(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisNavigationToggleReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        if (this.layout) {
            params.layout = this.layout;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "layout": ["layoutWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return navigationToggleCss; }
}, [0, "arcgis-navigation-toggle", {
        "referenceElement": [1, "reference-element"],
        "layout": [1537],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "toggle": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "layout": ["layoutWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-navigation-toggle"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-navigation-toggle":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisNavigationToggle$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisNavigationToggle = ArcgisNavigationToggle$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisNavigationToggle, defineCustomElement };
