// @arcgis/map-components/dist/components/arcgis-utility-network-validate-topology.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { a2 as newWidgetsUtilityNetworkValidateTopology } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const utilityNetworkValidateTopologyCss = "arcgis-utility-network-validate-topology{}";

const ArcgisUtilityNetworkValidateTopology$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisUtilityNetworkValidateTopology extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisUtilityNetworkValidateTopologyReady = createEvent(this, "arcgisUtilityNetworkValidateTopologyReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.extentToValidate = "current";
        this.utilityNetwork = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    extentToValidateWatcher(value) {
        if (this.widget) {
            this.widget.extentToValidate = value;
        }
    }
    utilityNetworkWatcher(value) {
        if (this.widget) {
            this.widget.utilityNetwork = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsUtilityNetworkValidateTopology(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisUtilityNetworkValidateTopologyReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        if (this.extentToValidate !== undefined) {
            params.extentToValidate = this.extentToValidate;
        }
        if (this.utilityNetwork !== undefined) {
            params.utilityNetwork = this.utilityNetwork;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "extentToValidate": ["extentToValidateWatcher"],
        "utilityNetwork": ["utilityNetworkWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return utilityNetworkValidateTopologyCss; }
}, [0, "arcgis-utility-network-validate-topology", {
        "referenceElement": [1, "reference-element"],
        "extentToValidate": [1, "extent-to-validate"],
        "utilityNetwork": [16],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "extentToValidate": ["extentToValidateWatcher"],
        "utilityNetwork": ["utilityNetworkWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-utility-network-validate-topology"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-utility-network-validate-topology":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisUtilityNetworkValidateTopology$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisUtilityNetworkValidateTopology = ArcgisUtilityNetworkValidateTopology$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisUtilityNetworkValidateTopology, defineCustomElement };
