// @arcgis/map-components/dist/components/arcgis-zoom.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { a6 as newWidgetsZoom } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const zoomCss = "arcgis-zoom{}";

const ArcgisZoom$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisZoom extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisZoomReady = createEvent(this, "arcgisZoomReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.icon = undefined;
        this.layout = "vertical";
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    iconWatcher(value) {
        if (this.widget && this.widget.icon !== value) {
            this.widget.icon = value;
        }
    }
    layoutWatcher(value) {
        if (this.widget && this.widget.layout !== value) {
            this.widget.layout = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async zoomIn() {
        this.widget?.viewModel.zoomIn();
    }
    async zoomOut() {
        this.widget?.viewModel.zoomOut();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsZoom(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisZoomReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.layout) {
            params.layout = this.layout;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "layout": ["layoutWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return zoomCss; }
}, [0, "arcgis-zoom", {
        "referenceElement": [1, "reference-element"],
        "icon": [1],
        "layout": [1],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "zoomIn": [64],
        "zoomOut": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "layout": ["layoutWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-zoom"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-zoom":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisZoom$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisZoom = ArcgisZoom$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisZoom, defineCustomElement };
