// @arcgis/map-components/dist/components/arcgis-utility-network-trace.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { a1 as newWidgetsUtilityNetworkTrace, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const utilityNetworkTraceCss = "";

const ArcgisUtilityNetworkTrace$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisUtilityNetworkTrace extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisUtilityNetworkTraceReady = createEvent(this, "arcgisUtilityNetworkTraceReady", 7);
        this.utilityNetworkTraceAddFlag = createEvent(this, "utilityNetworkTraceAddFlag", 7);
        this.utilityNetworkTraceAddFlagComplete = createEvent(this, "utilityNetworkTraceAddFlagComplete", 7);
        this.utilityNetworkTraceAddFlagError = createEvent(this, "utilityNetworkTraceAddFlagError", 7);
        this.utilityNetworkTraceAddResultArea = createEvent(this, "utilityNetworkTraceAddResultArea", 7);
        this.utilityNetworkTraceRemoveResultArea = createEvent(this, "utilityNetworkTraceRemoveResultArea", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.enableResultArea = false;
        this.defaultGraphicColor = {
            color: [255, 255, 0, 0.6],
            haloOpacity: 0.9,
            fillOpacity: 0.2,
            hex: "#FFFF00"
        };
        this.disabled = false;
        this.flags = [];
        this.gdbVersion = "sde.DEFAULT";
        this.inputSettings = [];
        this.resultAreaProperties = undefined;
        this.selectedTraces = [];
        this.selectOnComplete = true;
        this.showGraphicsOnComplete = true;
        this.showSelectionAttributes = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    enableResultAreaWatcher(value) {
        if (this.widget) {
            this.widget.enableResultArea = value;
        }
    }
    defaultGraphicColorWatcher(value) {
        if (this.widget) {
            this.widget.defaultGraphicColor = value;
        }
    }
    disabledWatcher(value, old) {
        if (this.widget && value !== old) {
            this.widget.disabled = value;
        }
    }
    flagsWatcher(value) {
        if (this.widget) {
            this.widget.flags = value;
        }
    }
    gdbVersionWatcher(value) {
        if (this.widget) {
            this.widget.gdbVersion = value;
        }
    }
    inputSettingsWatcher(value) {
        if (this.widget) {
            this.widget.inputSettings = value;
        }
    }
    resultAreaPropertiesWatcher(value) {
        if (this.widget) {
            this.widget.resultAreaProperties = value;
        }
    }
    selectedTracesWatcher(value) {
        if (this.widget) {
            this.widget.selectedTraces = value;
        }
    }
    selectOnCompleteWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.selectOnComplete = value;
        }
    }
    showGraphicsOnCompleteWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.showGraphicsOnComplete = value;
        }
    }
    showSelectionAttributesWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.showSelectionAttributes = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async checkCanTrace() {
        await this.widget?.checkCanTrace();
    }
    async confirmReset() {
        this.widget?.confirmReset();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsUtilityNetworkTrace(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisUtilityNetworkTraceReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.disabled, (value) => (this.disabled = value)));
    }
    _createParams() {
        const params = {};
        if (this.defaultGraphicColor) {
            params.defaultGraphicColor = this.defaultGraphicColor;
        }
        if (this.enableResultArea !== undefined) {
            params.enableResultArea = this.enableResultArea;
        }
        if (this.flags) {
            params.flags = this.flags;
        }
        if (this.gdbVersion) {
            params.gdbVersion = this.gdbVersion;
        }
        if (this.inputSettings) {
            params.inputSettings = this.inputSettings;
        }
        if (this.resultAreaProperties) {
            params.resultAreaProperties = this.resultAreaProperties;
        }
        if (this.selectedTraces) {
            params.selectedTraces = this.selectedTraces;
        }
        if (this.selectOnComplete !== undefined) {
            params.selectOnComplete = this.selectOnComplete;
        }
        if (this.showGraphicsOnComplete !== undefined) {
            params.showGraphicsOnComplete = this.showGraphicsOnComplete;
        }
        if (this.showSelectionAttributes !== undefined) {
            params.showSelectionAttributes = this.showSelectionAttributes;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    _proxyEvents() {
        this.widget.on("add-flag", (e) => this.utilityNetworkTraceAddFlag.emit(e));
        this.widget.on("add-flag-complete", (e) => this.utilityNetworkTraceAddFlagComplete.emit(e));
        this.widget.on("add-flag-error", (e) => this.utilityNetworkTraceAddFlagError.emit(e));
        this.widget.on("add-result-area", (e) => this.utilityNetworkTraceAddResultArea.emit(e));
        this.widget.on("remove-result-area", (e) => this.utilityNetworkTraceRemoveResultArea.emit(e));
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "enableResultArea": ["enableResultAreaWatcher"],
        "defaultGraphicColor": ["defaultGraphicColorWatcher"],
        "disabled": ["disabledWatcher"],
        "flags": ["flagsWatcher"],
        "gdbVersion": ["gdbVersionWatcher"],
        "inputSettings": ["inputSettingsWatcher"],
        "resultAreaProperties": ["resultAreaPropertiesWatcher"],
        "selectedTraces": ["selectedTracesWatcher"],
        "selectOnComplete": ["selectOnCompleteWatcher"],
        "showGraphicsOnComplete": ["showGraphicsOnCompleteWatcher"],
        "showSelectionAttributes": ["showSelectionAttributesWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return utilityNetworkTraceCss; }
}, [0, "arcgis-utility-network-trace", {
        "referenceElement": [1, "reference-element"],
        "enableResultArea": [4, "enable-result-area"],
        "defaultGraphicColor": [16],
        "disabled": [1540],
        "flags": [16],
        "gdbVersion": [1, "gdb-version"],
        "inputSettings": [16],
        "resultAreaProperties": [16],
        "selectedTraces": [16],
        "selectOnComplete": [4, "select-on-complete"],
        "showGraphicsOnComplete": [4, "show-graphics-on-complete"],
        "showSelectionAttributes": [4, "show-selection-attributes"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "checkCanTrace": [64],
        "confirmReset": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "enableResultArea": ["enableResultAreaWatcher"],
        "defaultGraphicColor": ["defaultGraphicColorWatcher"],
        "disabled": ["disabledWatcher"],
        "flags": ["flagsWatcher"],
        "gdbVersion": ["gdbVersionWatcher"],
        "inputSettings": ["inputSettingsWatcher"],
        "resultAreaProperties": ["resultAreaPropertiesWatcher"],
        "selectedTraces": ["selectedTracesWatcher"],
        "selectOnComplete": ["selectOnCompleteWatcher"],
        "showGraphicsOnComplete": ["showGraphicsOnCompleteWatcher"],
        "showSelectionAttributes": ["showSelectionAttributesWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-utility-network-trace"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-utility-network-trace":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisUtilityNetworkTrace$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisUtilityNetworkTrace = ArcgisUtilityNetworkTrace$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisUtilityNetworkTrace, defineCustomElement };
