// @arcgis/map-components/dist/components/arcgis-daylight.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { l as newWidgetsDaylight } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const daylightCss = "";

const ArcgisDaylight$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisDaylight extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisDaylightReady = createEvent(this, "arcgisDaylightReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.currentSeason = undefined;
        this.dayPlaying = false;
        this.yearPlaying = false;
        this.dateOrSeason = "date";
        this.icon = undefined;
        this.playSpeedMultiplier = 1.0;
        this.timeSliderSteps = 5;
        this.visibleElementsHeader = true;
        this.visibleElementsPlayButtons = true;
        this.visibleElementsDatePicker = true;
        this.visibleElementsTimezone = true;
        this.visibleElementsSunLightingToggle = true;
        this.visibleElementsShadowsToggle = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    currentSeasonWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.currentSeason = value;
        }
    }
    dayPlayingWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.dayPlaying = value;
        }
    }
    yearPlayingWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.yearPlaying = value;
        }
    }
    dateOrSeasonWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.dateOrSeason = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    playSpeedMultiplierWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.playSpeedMultiplier = value;
        }
    }
    timeSliderStepsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.timeSliderSteps = value;
        }
    }
    visibleElementsHeaderWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements.header = value;
        }
    }
    visibleElementsPlayButtonsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                playButtons: value
            };
        }
    }
    visibleElementsDatePickerWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                datePicker: value
            };
        }
    }
    visibleElementsTimezoneWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                timezone: value
            };
        }
    }
    visibleElementsSunLightingToggleWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                sunLightingToggle: value
            };
        }
    }
    visibleElementsShadowsToggleWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                shadowsToggle: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsDaylight(params);
        this.widget = widget;
        this._addContainer();
        this._setViewModelParams();
        this.arcgisDaylightReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            header: this.visibleElementsHeader,
            playButtons: this.visibleElementsPlayButtons,
            datePicker: this.visibleElementsDatePicker,
            timezone: this.visibleElementsTimezone,
            sunLightingToggle: this.visibleElementsSunLightingToggle,
            shadowsToggle: this.visibleElementsShadowsToggle
        };
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.view) {
            params.view = this.view;
        }
        if (this.dateOrSeason) {
            params.dateOrSeason = this.dateOrSeason;
        }
        if (this.playSpeedMultiplier) {
            params.playSpeedMultiplier = this.playSpeedMultiplier;
        }
        if (this.timeSliderSteps) {
            params.timeSliderSteps = this.timeSliderSteps;
        }
        return params;
    }
    _setViewModelParams() {
        const params = {
            dayPlaying: this.dayPlaying,
            yearPlaying: this.yearPlaying
        };
        if (this.currentSeason) {
            params.currentSeason = this.currentSeason;
        }
        this.widget.viewModel.set(params);
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "currentSeason": ["currentSeasonWatcher"],
        "dayPlaying": ["dayPlayingWatcher"],
        "yearPlaying": ["yearPlayingWatcher"],
        "dateOrSeason": ["dateOrSeasonWatcher"],
        "icon": ["iconWatcher"],
        "playSpeedMultiplier": ["playSpeedMultiplierWatcher"],
        "timeSliderSteps": ["timeSliderStepsWatcher"],
        "visibleElementsHeader": ["visibleElementsHeaderWatcher"],
        "visibleElementsPlayButtons": ["visibleElementsPlayButtonsWatcher"],
        "visibleElementsDatePicker": ["visibleElementsDatePickerWatcher"],
        "visibleElementsTimezone": ["visibleElementsTimezoneWatcher"],
        "visibleElementsSunLightingToggle": ["visibleElementsSunLightingToggleWatcher"],
        "visibleElementsShadowsToggle": ["visibleElementsShadowsToggleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return daylightCss; }
}, [0, "arcgis-daylight", {
        "referenceElement": [1, "reference-element"],
        "currentSeason": [1, "current-season"],
        "dayPlaying": [4, "day-playing"],
        "yearPlaying": [4, "year-playing"],
        "dateOrSeason": [1, "date-or-season"],
        "icon": [1],
        "playSpeedMultiplier": [2, "play-speed-multiplier"],
        "timeSliderSteps": [2, "time-slider-steps"],
        "visibleElementsHeader": [4, "visible-elements-header"],
        "visibleElementsPlayButtons": [4, "visible-elements-play-buttons"],
        "visibleElementsDatePicker": [4, "visible-elements-date-picker"],
        "visibleElementsTimezone": [4, "visible-elements-timezone"],
        "visibleElementsSunLightingToggle": [4, "visible-elements-sun-lighting-toggle"],
        "visibleElementsShadowsToggle": [4, "visible-elements-shadows-toggle"],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "currentSeason": ["currentSeasonWatcher"],
        "dayPlaying": ["dayPlayingWatcher"],
        "yearPlaying": ["yearPlayingWatcher"],
        "dateOrSeason": ["dateOrSeasonWatcher"],
        "icon": ["iconWatcher"],
        "playSpeedMultiplier": ["playSpeedMultiplierWatcher"],
        "timeSliderSteps": ["timeSliderStepsWatcher"],
        "visibleElementsHeader": ["visibleElementsHeaderWatcher"],
        "visibleElementsPlayButtons": ["visibleElementsPlayButtonsWatcher"],
        "visibleElementsDatePicker": ["visibleElementsDatePickerWatcher"],
        "visibleElementsTimezone": ["visibleElementsTimezoneWatcher"],
        "visibleElementsSunLightingToggle": ["visibleElementsSunLightingToggleWatcher"],
        "visibleElementsShadowsToggle": ["visibleElementsShadowsToggleWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-daylight"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-daylight":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisDaylight$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisDaylight = ArcgisDaylight$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisDaylight, defineCustomElement };
