// @arcgis/map-components/dist/components/arcgis-histogram.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { C as newWidgetsHistogram, i as importCoreReactiveUtils } from './index2.js';
import { a as addHandles } from './component-utils.js';
import { d as defineCustomElement$2 } from './expand.js';

const histogramCss = "arcgis-histogram{}";

const ArcgisHistogram$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisHistogram extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.widgetReady = createEvent(this, "widgetReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.average = undefined;
        this.barCreatedFunction = undefined;
        this.bins = undefined;
        this.dataLineCreatedFunction = undefined;
        this.dataLines = undefined;
        this.labelFormatFunction = undefined;
        this.layout = undefined;
        this.max = 0;
        this.min = 0;
        this.state = undefined;
        this.position = "bottom-left";
        this.expandable = false;
        this.widget = undefined;
    }
    averageWatcher(newValue) {
        if (this.widget) {
            this.widget.average = newValue;
        }
    }
    barCreatedFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.barCreatedFunction = newValue;
        }
    }
    binsWatcher(newValue) {
        if (this.widget) {
            this.widget.bins = newValue;
        }
    }
    dataLineCreatedFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.dataLineCreatedFunction = newValue;
        }
    }
    dataLinesWatcher(newValue) {
        if (this.widget) {
            this.widget.dataLines = newValue;
        }
    }
    labelFormatFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.labelFormatFunction = newValue;
        }
    }
    layoutWatcher(newValue) {
        if (this.widget) {
            this.widget.layout = newValue;
        }
    }
    maxWatcher(newValue) {
        if (this.widget) {
            this.widget.max = Number(newValue);
        }
    }
    minWatcher(newValue) {
        if (this.widget) {
            this.widget.min = Number(newValue);
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsHistogram(params);
        this.widget = widget;
        await this._addWatchers();
        this.widgetReady.emit({ widget: this.widget });
    }
    render() {
        if (this.expandable) {
            return (h("arcgis-expand", { position: this.position, ref: (el) => (this.expandElem = el) }, h("div", { ref: (el) => (this.childElem = el) })));
        }
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.state, () => (this.state = this.widget.state), {
            initial: true
        }));
    }
    _createParams() {
        const params = {
            container: this.childElem,
            min: this.min,
            max: this.max
        };
        if (this.average !== undefined) {
            params.average = this.average;
        }
        if (this.barCreatedFunction) {
            params.barCreatedFunction = this.barCreatedFunction;
        }
        if (this.bins !== undefined) {
            params.bins = this.bins;
        }
        if (this.dataLineCreatedFunction) {
            params.dataLineCreatedFunction = this.dataLineCreatedFunction;
        }
        if (this.dataLines !== undefined) {
            params.dataLines = this.dataLines;
        }
        if (this.labelFormatFunction) {
            params.labelFormatFunction = this.labelFormatFunction;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "average": ["averageWatcher"],
        "barCreatedFunction": ["barCreatedFunctionWatcher"],
        "bins": ["binsWatcher"],
        "dataLineCreatedFunction": ["dataLineCreatedFunctionWatcher"],
        "dataLines": ["dataLinesWatcher"],
        "labelFormatFunction": ["labelFormatFunctionWatcher"],
        "layout": ["layoutWatcher"],
        "max": ["maxWatcher"],
        "min": ["minWatcher"]
    }; }
    static get style() { return histogramCss; }
}, [0, "arcgis-histogram", {
        "referenceElement": [1, "reference-element"],
        "average": [2],
        "barCreatedFunction": [16],
        "bins": [16],
        "dataLineCreatedFunction": [16],
        "dataLines": [16],
        "labelFormatFunction": [16],
        "layout": [1],
        "max": [2],
        "min": [2],
        "state": [1025],
        "position": [1],
        "expandable": [4],
        "widget": [1040]
    }, undefined, {
        "average": ["averageWatcher"],
        "barCreatedFunction": ["barCreatedFunctionWatcher"],
        "bins": ["binsWatcher"],
        "dataLineCreatedFunction": ["dataLineCreatedFunctionWatcher"],
        "dataLines": ["dataLinesWatcher"],
        "labelFormatFunction": ["labelFormatFunctionWatcher"],
        "layout": ["layoutWatcher"],
        "max": ["maxWatcher"],
        "min": ["minWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-histogram", "arcgis-expand"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-histogram":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisHistogram$1);
            }
            break;
        case "arcgis-expand":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisHistogram = ArcgisHistogram$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisHistogram, defineCustomElement };
