// @arcgis/map-components/dist/components/arcgis-feature.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { u as newWidgetsFeature, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const featureCss = "";

const ArcgisFeature$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisFeature extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisFeatureReady = createEvent(this, "arcgisFeatureReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.defaultPopupTemplateEnabled = false;
        this.graphic = undefined;
        this.location = undefined;
        this.map = undefined;
        this.spatialReference = undefined;
        this.visibleElementsTitle = true;
        this.visibleElementsContentAttachments = true;
        this.visibleElementsContentFields = true;
        this.visibleElementsContentMedia = true;
        this.visibleElementsContentText = true;
        this.visibleElementsContentExpression = true;
        this.visibleElementsLastEditInfo = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    defaultPopupTemplateEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.defaultPopupTemplateEnabled = value;
        }
    }
    graphicWatcher(value) {
        if (this.widget) {
            this.widget.graphic = value;
        }
    }
    locationWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.location = value;
        }
    }
    spatialReferenceWatcher(value) {
        if (this.widget) {
            this.widget.spatialReference = value;
        }
    }
    visibleElementsTitleWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                title: value
            };
        }
    }
    visibleElementsContentAttachmentsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                content: {
                    ...this.widget.visibleElements.content,
                    attachments: value
                }
            };
        }
    }
    visibleElementsContentFieldsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                content: {
                    ...this.widget.visibleElements.content,
                    fields: value
                }
            };
        }
    }
    visibleElementsContentMediaWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                content: {
                    ...this.widget.visibleElements.content,
                    media: value
                }
            };
        }
    }
    visibleElementsContentTextWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                content: {
                    ...this.widget.visibleElements.content,
                    text: value
                }
            };
        }
    }
    visibleElementsContentExpressionWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                content: {
                    ...this.widget.visibleElements.content,
                    expression: value
                }
            };
        }
    }
    visibleElementsLastEditInfoWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                lastEditInfo: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async nextMedia(contentElementIndex) {
        this.widget?.nextMedia(contentElementIndex);
    }
    async previousMedia(contentElementIndex) {
        this.widget?.previousMedia(contentElementIndex);
    }
    async setActiveMedia(contentElementIndex, mediaInfoIndex) {
        this.widget?.setActiveMedia(contentElementIndex, mediaInfoIndex);
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsFeature(params);
        this.widget = widget;
        this._addContainer();
        this._setViewModelParams();
        await this._addWatchers();
        this.arcgisFeatureReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.view?.map, (value) => (this.map = value)));
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            title: this.visibleElementsTitle,
            content: {
                attachments: this.visibleElementsContentAttachments,
                fields: this.visibleElementsContentFields,
                media: this.visibleElementsContentMedia,
                text: this.visibleElementsContentText,
                expression: this.visibleElementsContentExpression
            },
            lastEditInfo: this.visibleElementsLastEditInfo
        };
        if (this.defaultPopupTemplateEnabled !== undefined) {
            params.defaultPopupTemplateEnabled = this.defaultPopupTemplateEnabled;
        }
        if (this.graphic) {
            params.graphic = this.graphic;
        }
        if (this.view) {
            params.view = this.view;
        }
        if (this.spatialReference) {
            params.spatialReference = this.spatialReference;
        }
        return params;
    }
    _setViewModelParams() {
        const params = {};
        if (this.location) {
            params.location = this.location;
        }
        this.widget.viewModel.set(params);
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "defaultPopupTemplateEnabled": ["defaultPopupTemplateEnabledWatcher"],
        "graphic": ["graphicWatcher"],
        "location": ["locationWatcher"],
        "spatialReference": ["spatialReferenceWatcher"],
        "visibleElementsTitle": ["visibleElementsTitleWatcher"],
        "visibleElementsContentAttachments": ["visibleElementsContentAttachmentsWatcher"],
        "visibleElementsContentFields": ["visibleElementsContentFieldsWatcher"],
        "visibleElementsContentMedia": ["visibleElementsContentMediaWatcher"],
        "visibleElementsContentText": ["visibleElementsContentTextWatcher"],
        "visibleElementsContentExpression": ["visibleElementsContentExpressionWatcher"],
        "visibleElementsLastEditInfo": ["visibleElementsLastEditInfoWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return featureCss; }
}, [0, "arcgis-feature", {
        "referenceElement": [1, "reference-element"],
        "defaultPopupTemplateEnabled": [4, "default-popup-template-enabled"],
        "graphic": [16],
        "location": [16],
        "map": [1040],
        "spatialReference": [16],
        "visibleElementsTitle": [4, "visible-elements-title"],
        "visibleElementsContentAttachments": [4, "visible-elements-content-attachments"],
        "visibleElementsContentFields": [4, "visible-elements-content-fields"],
        "visibleElementsContentMedia": [4, "visible-elements-content-media"],
        "visibleElementsContentText": [4, "visible-elements-content-text"],
        "visibleElementsContentExpression": [4, "visible-elements-content-expression"],
        "visibleElementsLastEditInfo": [4, "visible-elements-last-edit-info"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "nextMedia": [64],
        "previousMedia": [64],
        "setActiveMedia": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "defaultPopupTemplateEnabled": ["defaultPopupTemplateEnabledWatcher"],
        "graphic": ["graphicWatcher"],
        "location": ["locationWatcher"],
        "spatialReference": ["spatialReferenceWatcher"],
        "visibleElementsTitle": ["visibleElementsTitleWatcher"],
        "visibleElementsContentAttachments": ["visibleElementsContentAttachmentsWatcher"],
        "visibleElementsContentFields": ["visibleElementsContentFieldsWatcher"],
        "visibleElementsContentMedia": ["visibleElementsContentMediaWatcher"],
        "visibleElementsContentText": ["visibleElementsContentTextWatcher"],
        "visibleElementsContentExpression": ["visibleElementsContentExpressionWatcher"],
        "visibleElementsLastEditInfo": ["visibleElementsLastEditInfoWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-feature"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-feature":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisFeature$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisFeature = ArcgisFeature$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisFeature, defineCustomElement };
