// @arcgis/map-components/dist/components/arcgis-fullscreen.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { B as newWidgetsFullscreen } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const fullscreenCss = "";

const ArcgisFullscreen$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisFullscreen extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.widgetReady = createEvent(this, "widgetReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.element = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(newValue) {
        referenceElementHelper(this, newValue);
    }
    elementWatcher(newValue) {
        if (this.widget) {
            this.widget.element = newValue;
        }
    }
    positionWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.view.ui.move(this.childElem, newValue);
        }
    }
    viewWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.view = newValue;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsFullscreen(params);
        this.widget = widget;
        this.widgetReady.emit({ widget: this.widget });
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _createParams() {
        const params = {
            container: this.childElem
        };
        if (this.element) {
            params.element = this.element;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "element": ["elementWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return fullscreenCss; }
}, [0, "arcgis-fullscreen", {
        "referenceElement": [1, "reference-element"],
        "element": [16],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "element": ["elementWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-fullscreen"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-fullscreen":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisFullscreen$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisFullscreen = ArcgisFullscreen$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisFullscreen, defineCustomElement };
