// @arcgis/map-components/dist/components/arcgis-weather.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { a5 as newWidgetsWeather } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const weatherCss = "";

const ArcgisWeather$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisWeather extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisWeatherReady = createEvent(this, "arcgisWeatherReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.icon = undefined;
        this.visibleElementsHeader = true;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    visibleElementsHeaderWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                header: value
            };
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsWeather(params);
        this.widget = widget;
        this._addContainer();
        this.arcgisWeatherReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            header: this.visibleElementsHeader
        };
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "visibleElementsHeader": ["visibleElementsHeaderWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return weatherCss; }
}, [0, "arcgis-weather", {
        "referenceElement": [1, "reference-element"],
        "icon": [1],
        "visibleElementsHeader": [4, "visible-elements-header"],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "visibleElementsHeader": ["visibleElementsHeaderWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-weather"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-weather":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisWeather$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisWeather = ArcgisWeather$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisWeather, defineCustomElement };
