// @arcgis/map-components/dist/components/arcgis-table-list.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { Z as newWidgetsTableList, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const tableListCss = "arcgis-table-list{}";

const ArcgisTableList$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisTableList extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisTableListReady = createEvent(this, "arcgisTableListReady", 7);
        this.tableListTriggerAction = createEvent(this, "tableListTriggerAction", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.collapsed = false;
        this.dragEnabled = false;
        this.filterPlaceholder = "";
        this.filterText = "";
        this.icon = undefined;
        this.minDragEnabledItems = 2;
        this.minFilterItems = 10;
        this.selectedItems = undefined;
        this.selectionMode = "none";
        this.listItemCreatedFunction = undefined;
        this.visibleElementsCloseButton = false;
        this.visibleElementsCollapseButton = false;
        this.visibleElementsErrors = false;
        this.visibleElementsFilter = false;
        this.visibleElementsFlow = true;
        this.visibleElementsHeading = false;
        this.visibleElementsStatusIndicators = true;
        this.visibleElementsTemporaryTableIndicators = false;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(newValue) {
        void referenceElementHelper(this, newValue);
    }
    collapsedWatcher(value) {
        if (this.widget && this.widget.collapsed !== value) {
            this.widget.collapsed = value;
        }
    }
    dragEnabledWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.dragEnabled = value;
        }
    }
    filterPlaceholderWatcher(value) {
        if (this.widget) {
            this.widget.filterPlaceholder = value;
        }
    }
    filterTextWatcher(value) {
        if (this.widget) {
            this.widget.filterText = value;
        }
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    minDragEnabledItemsWatcher(value) {
        if (this.widget) {
            this.widget.minDragEnabledItems = value;
        }
    }
    minFilterItemsWatcher(value) {
        if (this.widget) {
            this.widget.minFilterItems = value;
        }
    }
    selectedItemsWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.selectedItems = value;
        }
    }
    selectionModeWatcher(value) {
        if (this.widget) {
            this.widget.selectionMode = value;
        }
    }
    listItemCreatedFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.listItemCreatedFunction = newValue;
        }
    }
    visibleElementsCloseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                closeButton: value
            };
        }
    }
    visibleElementsCollapseButtonWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                collapseButton: value
            };
        }
    }
    visibleElementsErrorsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                errors: value
            };
        }
    }
    visibleElementsFilterWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                filter: value
            };
        }
    }
    visibleElementsFlowWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                flow: value
            };
        }
    }
    visibleElementsHeadingWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                heading: value
            };
        }
    }
    visibleElementsStatusIndicatorsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                statusIndicators: value
            };
        }
    }
    visibleElementsTemporaryTableIndicatorsWatcher(value) {
        if (this.widget) {
            this.widget.visibleElements = {
                ...this.widget.visibleElements,
                temporaryTableIndicators: value
            };
        }
    }
    positionWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.view.ui.move(this.childElem, newValue);
        }
    }
    viewWatcher(newValue) {
        if (newValue !== undefined && this.widget) {
            this.widget.map = newValue.map;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsTableList(params);
        this.widget = widget;
        await this._addWatchers();
        this._proxyEvents();
        this._addContainer();
        this.arcgisTableListReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.selectedItems, (value) => (this.selectedItems = value)),
            watch(() => this.widget.collapsed, (value) => (this.collapsed = value))
        ]);
    }
    _createParams() {
        const params = {};
        params.visibleElements = {
            closeButton: this.visibleElementsCloseButton,
            collapseButton: this.visibleElementsCollapseButton,
            errors: this.visibleElementsErrors,
            filter: this.visibleElementsFilter,
            flow: this.visibleElementsFlow,
            heading: this.visibleElementsHeading,
            statusIndicators: this.visibleElementsStatusIndicators,
            temporaryTableIndicators: this.visibleElementsTemporaryTableIndicators
        };
        if (this.collapsed !== undefined) {
            params.collapsed = this.collapsed;
        }
        if (this.dragEnabled !== undefined) {
            params.dragEnabled = this.dragEnabled;
        }
        if (this.filterPlaceholder) {
            params.filterPlaceholder = this.filterPlaceholder;
        }
        if (this.filterText) {
            params.filterText = this.filterText;
        }
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.listItemCreatedFunction) {
            params.listItemCreatedFunction = this.listItemCreatedFunction;
        }
        if (this.minDragEnabledItems !== undefined) {
            params.minDragEnabledItems = this.minDragEnabledItems;
        }
        if (this.minFilterItems !== undefined) {
            params.minFilterItems = this.minFilterItems;
        }
        if (this.selectedItems) {
            params.selectedItems = this.selectedItems;
        }
        if (this.selectionMode) {
            params.selectionMode = this.selectionMode;
        }
        if (this.view) {
            params.map = this.view.map;
        }
        return params;
    }
    _proxyEvents() {
        this.widget?.on("trigger-action", (e) => this.tableListTriggerAction.emit(e));
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "collapsed": ["collapsedWatcher"],
        "dragEnabled": ["dragEnabledWatcher"],
        "filterPlaceholder": ["filterPlaceholderWatcher"],
        "filterText": ["filterTextWatcher"],
        "icon": ["iconWatcher"],
        "minDragEnabledItems": ["minDragEnabledItemsWatcher"],
        "minFilterItems": ["minFilterItemsWatcher"],
        "selectedItems": ["selectedItemsWatcher"],
        "selectionMode": ["selectionModeWatcher"],
        "listItemCreatedFunction": ["listItemCreatedFunctionWatcher"],
        "visibleElementsCloseButton": ["visibleElementsCloseButtonWatcher"],
        "visibleElementsCollapseButton": ["visibleElementsCollapseButtonWatcher"],
        "visibleElementsErrors": ["visibleElementsErrorsWatcher"],
        "visibleElementsFilter": ["visibleElementsFilterWatcher"],
        "visibleElementsFlow": ["visibleElementsFlowWatcher"],
        "visibleElementsHeading": ["visibleElementsHeadingWatcher"],
        "visibleElementsStatusIndicators": ["visibleElementsStatusIndicatorsWatcher"],
        "visibleElementsTemporaryTableIndicators": ["visibleElementsTemporaryTableIndicatorsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return tableListCss; }
}, [0, "arcgis-table-list", {
        "referenceElement": [1, "reference-element"],
        "collapsed": [1540],
        "dragEnabled": [4, "drag-enabled"],
        "filterPlaceholder": [1, "filter-placeholder"],
        "filterText": [1, "filter-text"],
        "icon": [1],
        "minDragEnabledItems": [2, "min-drag-enabled-items"],
        "minFilterItems": [2, "min-filter-items"],
        "selectedItems": [1040],
        "selectionMode": [1, "selection-mode"],
        "listItemCreatedFunction": [16],
        "visibleElementsCloseButton": [4, "visible-elements-close-button"],
        "visibleElementsCollapseButton": [4, "visible-elements-collapse-button"],
        "visibleElementsErrors": [4, "visible-elements-errors"],
        "visibleElementsFilter": [4, "visible-elements-filter"],
        "visibleElementsFlow": [4, "visible-elements-flow"],
        "visibleElementsHeading": [4, "visible-elements-heading"],
        "visibleElementsStatusIndicators": [4, "visible-elements-status-indicators"],
        "visibleElementsTemporaryTableIndicators": [4, "visible-elements-temporary-table-indicators"],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "collapsed": ["collapsedWatcher"],
        "dragEnabled": ["dragEnabledWatcher"],
        "filterPlaceholder": ["filterPlaceholderWatcher"],
        "filterText": ["filterTextWatcher"],
        "icon": ["iconWatcher"],
        "minDragEnabledItems": ["minDragEnabledItemsWatcher"],
        "minFilterItems": ["minFilterItemsWatcher"],
        "selectedItems": ["selectedItemsWatcher"],
        "selectionMode": ["selectionModeWatcher"],
        "listItemCreatedFunction": ["listItemCreatedFunctionWatcher"],
        "visibleElementsCloseButton": ["visibleElementsCloseButtonWatcher"],
        "visibleElementsCollapseButton": ["visibleElementsCollapseButtonWatcher"],
        "visibleElementsErrors": ["visibleElementsErrorsWatcher"],
        "visibleElementsFilter": ["visibleElementsFilterWatcher"],
        "visibleElementsFlow": ["visibleElementsFlowWatcher"],
        "visibleElementsHeading": ["visibleElementsHeadingWatcher"],
        "visibleElementsStatusIndicators": ["visibleElementsStatusIndicatorsWatcher"],
        "visibleElementsTemporaryTableIndicators": ["visibleElementsTemporaryTableIndicatorsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-table-list"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-table-list":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisTableList$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisTableList = ArcgisTableList$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisTableList, defineCustomElement };
