// @arcgis/map-components/dist/components/arcgis-slice.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { W as newWidgetsSlice, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const sliceCss = "arcgis-slice{}";

const ArcgisSlice$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisSlice extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisSliceReady = createEvent(this, "arcgisSliceReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.analysis = undefined;
        this.excludedLayers = undefined;
        this.excludeGroundSurface = false;
        this.icon = undefined;
        this.shape = undefined;
        this.tiltEnabled = false;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    excludedLayersWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.excludedLayers = value;
        }
    }
    excludeGroundSurfaceWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.excludeGroundSurface = value;
        }
    }
    iconWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.icon = value;
        }
    }
    shapeWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.shape = value;
        }
    }
    tiltEnabledWatcher(value) {
        if (this.widget) {
            this.widget.viewModel.tiltEnabled = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async clear() {
        this.widget?.viewModel.clear();
    }
    async start() {
        this.widget?.viewModel.start();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsSlice(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisSliceReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.analysis, (value) => {
            this.analysis = value;
            this.arcgisPropertyChange.emit({ name: "analysis" });
        }, {
            initial: true
        }));
    }
    _createParams() {
        const params = {};
        params.viewModel = {
            excludeGroundSurface: this.excludeGroundSurface,
            tiltEnabled: this.tiltEnabled
        };
        if (this.excludedLayers) {
            params.viewModel.excludedLayers = this.excludedLayers;
        }
        if (this.shape) {
            params.viewModel.shape = this.shape;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "excludedLayers": ["excludedLayersWatcher"],
        "excludeGroundSurface": ["excludeGroundSurfaceWatcher"],
        "icon": ["iconWatcher"],
        "shape": ["shapeWatcher"],
        "tiltEnabled": ["tiltEnabledWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return sliceCss; }
}, [0, "arcgis-slice", {
        "referenceElement": [1, "reference-element"],
        "analysis": [1040],
        "excludedLayers": [16],
        "excludeGroundSurface": [4, "exclude-ground-surface"],
        "icon": [1],
        "shape": [16],
        "tiltEnabled": [4, "tilt-enabled"],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "clear": [64],
        "start": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "excludedLayers": ["excludedLayersWatcher"],
        "excludeGroundSurface": ["excludeGroundSurfaceWatcher"],
        "icon": ["iconWatcher"],
        "shape": ["shapeWatcher"],
        "tiltEnabled": ["tiltEnabledWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-slice"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-slice":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisSlice$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisSlice = ArcgisSlice$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisSlice, defineCustomElement };
