// @arcgis/map-components/dist/components/arcgis-floor-filter.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { A as newWidgetsFloorFilter, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const floorFilterCss = "";

const ArcgisFloorFilter$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisFloorFilter extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisFloorFilterReady = createEvent(this, "arcgisFloorFilterReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.facility = undefined;
        this.level = undefined;
        this.longNames = true;
        this.site = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    facilityWatcher(value) {
        if (this.widget) {
            this.widget.facility = value;
        }
    }
    levelWatcher(value) {
        if (this.widget) {
            this.widget.level = value;
        }
    }
    longNamesWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.longNames = value;
        }
    }
    siteWatcher(value) {
        if (this.widget) {
            this.widget.site = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async updateWebDocument(webmap) {
        this.widget?.updateWebDocument(webmap);
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsFloorFilter(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisFloorFilterReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch } = await importCoreReactiveUtils();
        addHandles(this, [
            watch(() => this.widget.facility, (value) => (this.facility = value)),
            watch(() => this.widget.level, (value) => (this.level = value)),
            watch(() => this.widget.site, (value) => (this.site = value))
        ]);
    }
    _createParams() {
        const params = {};
        if (this.facility) {
            params.facility = this.facility;
        }
        if (this.level) {
            params.level = this.level;
        }
        if (this.longNames !== undefined) {
            params.longNames = this.longNames;
        }
        if (this.site) {
            params.site = this.site;
        }
        if (this.view) {
            params.view = this.view;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "facility": ["facilityWatcher"],
        "level": ["levelWatcher"],
        "longNames": ["longNamesWatcher"],
        "site": ["siteWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return floorFilterCss; }
}, [0, "arcgis-floor-filter", {
        "referenceElement": [1, "reference-element"],
        "facility": [1537],
        "level": [1537],
        "longNames": [4, "long-names"],
        "site": [1537],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "updateWebDocument": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "facility": ["facilityWatcher"],
        "level": ["levelWatcher"],
        "longNames": ["longNamesWatcher"],
        "site": ["siteWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-floor-filter"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-floor-filter":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisFloorFilter$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisFloorFilter = ArcgisFloorFilter$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisFloorFilter, defineCustomElement };
