// @arcgis/map-components/dist/components/arcgis-distance-measurement-2d.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { r as newWidgetsDistanceMeasurement2D, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const distanceMeasurement2dCss = "";

const ArcgisDistanceMeasurement2d$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisDistanceMeasurement2d extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisDistanceMeasurement2dReady = createEvent(this, "arcgisDistanceMeasurement2dReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.icon = undefined;
        this.unit = undefined;
        this.unitOptions = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    unitWatcher(value) {
        if (this.widget) {
            this.widget.unit = value;
        }
    }
    unitOptionsWatcher(value) {
        if (this.widget) {
            this.widget.unitOptions = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsDistanceMeasurement2D(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisDistanceMeasurement2dReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { whenOnce } = await importCoreReactiveUtils();
        await whenOnce(() => this.widget.unitOptions);
        this.unitOptions = this.widget.unitOptions;
    }
    _createParams() {
        const params = {};
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.view) {
            params.view = this.view;
        }
        if (this.unit) {
            params.unit = this.unit;
        }
        if (this.unitOptions) {
            params.unitOptions = this.unitOptions;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "unit": ["unitWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return distanceMeasurement2dCss; }
}, [0, "arcgis-distance-measurement-2d", {
        "referenceElement": [1, "reference-element"],
        "icon": [1],
        "unit": [1],
        "unitOptions": [16],
        "position": [1],
        "view": [1040],
        "widget": [1040]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "unit": ["unitWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-distance-measurement-2d"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-distance-measurement-2d":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisDistanceMeasurement2d$1);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisDistanceMeasurement2d = ArcgisDistanceMeasurement2d$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisDistanceMeasurement2d, defineCustomElement };
