// @arcgis/map-components/dist/components/arcgis-directline-measurement-3d.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { q as newWidgetsDirectLineMeasurement3D, i as importCoreReactiveUtils } from './index2.js';
import { r as referenceElementHelper, c as checkForView, a as addHandles } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';

const directlineMeasurement3dCss = "";

const ArcgisDirectLineMeasurement3d = /*@__PURE__*/ proxyCustomElement(class ArcgisDirectLineMeasurement3d extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.arcgisDirectLineMeasurement3dReady = createEvent(this, "arcgisDirectLineMeasurement3dReady", 7);
        this.arcgisPropertyChange = createEvent(this, "arcgisPropertyChange", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.analysis = undefined;
        this.icon = undefined;
        this.unit = undefined;
        this.unitOptions = undefined;
        this.position = "bottom-left";
        this.view = undefined;
        this.widget = undefined;
    }
    referenceElementWatcher(value) {
        void referenceElementHelper(this, value);
    }
    iconWatcher(value) {
        if (this.widget) {
            this.widget.icon = value;
        }
    }
    unitWatcher(value) {
        if (this.widget) {
            this.widget.unit = value;
        }
    }
    unitOptionsWatcher(value) {
        if (this.widget) {
            this.widget.unitOptions = value;
        }
    }
    positionWatcher(value) {
        if (value !== undefined && this.widget) {
            this.view.ui.move(this.childElem, value);
        }
    }
    viewWatcher(value) {
        if (value !== undefined && this.widget) {
            this.widget.view = value;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Methods
    //
    //--------------------------------------------------------------------------
    async clear() {
        this.widget?.viewModel.clear();
    }
    async start() {
        this.widget?.viewModel.start();
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsDirectLineMeasurement3D(params);
        this.widget = widget;
        await this._addWatchers();
        this._addContainer();
        this.arcgisDirectLineMeasurement3dReady.emit();
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _addContainer() {
        this.widget.container = this.childElem;
    }
    async _addWatchers() {
        const { watch, whenOnce } = await importCoreReactiveUtils();
        addHandles(this, watch(() => this.widget.analysis, (value) => {
            this.analysis = value;
            this.arcgisPropertyChange.emit({ name: "analysis" });
        }, {
            initial: true
        }));
        await whenOnce(() => this.widget.unitOptions);
        this.unitOptions = this.widget.unitOptions;
    }
    _createParams() {
        const params = {};
        if (this.icon) {
            params.icon = this.icon;
        }
        if (this.view) {
            params.view = this.view;
        }
        if (this.unit) {
            params.unit = this.unit;
        }
        if (this.unitOptions) {
            params.unitOptions = this.unitOptions;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "unit": ["unitWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }; }
    static get style() { return directlineMeasurement3dCss; }
}, [0, "arcgis-directline-measurement-3d", {
        "referenceElement": [1, "reference-element"],
        "analysis": [1040],
        "icon": [1],
        "unit": [1],
        "unitOptions": [16],
        "position": [1],
        "view": [1040],
        "widget": [1040],
        "clear": [64],
        "start": [64]
    }, undefined, {
        "referenceElement": ["referenceElementWatcher"],
        "icon": ["iconWatcher"],
        "unit": ["unitWatcher"],
        "unitOptions": ["unitOptionsWatcher"],
        "position": ["positionWatcher"],
        "view": ["viewWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-directline-measurement-3d"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-directline-measurement-3d":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisDirectLineMeasurement3d);
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisDirectlineMeasurement3d = ArcgisDirectLineMeasurement3d;
const defineCustomElement = defineCustomElement$1;

export { ArcgisDirectlineMeasurement3d, defineCustomElement };
