// @arcgis/map-components/dist/components/arcgis-histogram-range-slider.js
/*!
 * All material copyright Esri, All Rights Reserved, unless otherwise specified.
 * See https://js.arcgis.com/4.30/esri/copyright.txt for details.
 * v4.30.0-next.13
 */
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { D as newWidgetsHistogramRangeSlider } from './index2.js';
import { c as checkForView } from './component-utils.js';
import { o as onDisconnectedHelper } from './disconnected.js';
import { d as defineCustomElement$2 } from './expand.js';

const histogramRangeSliderCss = "arcgis-histogram-range-slider{}";

const ArcgisHistogramRangeSlider$1 = /*@__PURE__*/ proxyCustomElement(class ArcgisHistogramRangeSlider extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.widgetReady = createEvent(this, "widgetReady", 7);
        this._watchHandles = [];
        this.referenceElement = undefined;
        this.average = undefined;
        this.barCreatedFunction = undefined;
        this.bins = undefined;
        this.dataLineCreatedFunction = undefined;
        this.dataLines = undefined;
        this.excludedBarColor = undefined;
        this.includedBarColor = undefined;
        this.labelFormatFunction = undefined;
        this.max = 0;
        this.min = 0;
        this.precision = undefined;
        this.rangeType = undefined;
        this.standardDeviation = undefined;
        this.standardDeviationCount = undefined;
        this.values = undefined;
        this.position = "bottom-left";
        this.expandable = false;
        this.widget = undefined;
    }
    averageWatcher(newValue) {
        if (this.widget) {
            this.widget.average = newValue;
        }
    }
    barCreatedFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.barCreatedFunction = newValue;
        }
    }
    binsWatcher(newValue) {
        if (this.widget) {
            this.widget.bins = newValue;
        }
    }
    dataLineCreatedFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.dataLineCreatedFunction = newValue;
        }
    }
    dataLinesWatcher(newValue) {
        if (this.widget) {
            this.widget.dataLines = newValue;
        }
    }
    excludedBarColorWatcher(newValue) {
        if (this.widget) {
            this.widget.excludedBarColor = newValue;
        }
    }
    includedBarColorWatcher(newValue) {
        if (this.widget) {
            this.widget.includedBarColor = newValue;
        }
    }
    labelFormatFunctionWatcher(newValue) {
        if (this.widget) {
            this.widget.labelFormatFunction = newValue;
        }
    }
    maxWatcher(newValue) {
        if (this.widget) {
            this.widget.max = Number(newValue);
        }
    }
    minWatcher(newValue) {
        if (this.widget) {
            this.widget.min = Number(newValue);
        }
    }
    precisionWatcher(newValue) {
        if (this.widget) {
            this.widget.precision = newValue;
        }
    }
    rangeTypeWatcher(newValue) {
        if (this.widget) {
            this.widget.rangeType = newValue;
        }
    }
    standardDeviationWatcher(newValue) {
        if (this.widget) {
            this.widget.standardDeviation = newValue;
        }
    }
    standardDeviationCountWatcher(newValue) {
        if (this.widget) {
            this.widget.standardDeviationCount = newValue;
        }
    }
    valuesWatcher(newValue) {
        if (this.widget) {
            this.widget.values = newValue;
        }
    }
    //--------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    //--------------------------------------------------------------------------
    async componentDidLoad() {
        const params = this._createParams();
        const widget = await newWidgetsHistogramRangeSlider(params);
        this.widget = widget;
        this.widgetReady.emit({ widget: this.widget });
        checkForView(this);
    }
    disconnectedCallback() {
        onDisconnectedHelper(this);
    }
    render() {
        if (this.expandable) {
            return (h("arcgis-expand", { position: this.position, ref: (el) => (this.expandElem = el) }, h("div", { ref: (el) => (this.childElem = el) })));
        }
        return h("div", { ref: (el) => (this.childElem = el) });
    }
    //--------------------------------------------------------------------------
    //
    //  Private Methods
    //
    //--------------------------------------------------------------------------
    _createParams() {
        const params = {
            container: this.childElem,
            min: this.min,
            max: this.max
        };
        if (this.average !== undefined) {
            params.average = this.average;
        }
        if (this.barCreatedFunction) {
            params.barCreatedFunction = this.barCreatedFunction;
        }
        if (this.bins !== undefined) {
            params.bins = this.bins;
        }
        if (this.dataLineCreatedFunction) {
            params.dataLineCreatedFunction = this.dataLineCreatedFunction;
        }
        if (this.dataLines !== undefined) {
            params.dataLines = this.dataLines;
        }
        if (this.excludedBarColor) {
            params.excludedBarColor = this.excludedBarColor;
        }
        if (this.includedBarColor) {
            params.includedBarColor = this.includedBarColor;
        }
        if (this.labelFormatFunction) {
            params.labelFormatFunction = this.labelFormatFunction;
        }
        if (this.precision !== undefined) {
            params.precision = this.precision;
        }
        if (this.rangeType) {
            params.rangeType = this.rangeType;
        }
        if (this.standardDeviation !== undefined) {
            params.standardDeviation = this.standardDeviation;
        }
        if (this.standardDeviationCount !== undefined) {
            params.standardDeviationCount = this.standardDeviationCount;
        }
        if (this.values !== undefined) {
            params.values = this.values;
        }
        return params;
    }
    get el() { return this; }
    static get watchers() { return {
        "average": ["averageWatcher"],
        "barCreatedFunction": ["barCreatedFunctionWatcher"],
        "bins": ["binsWatcher"],
        "dataLineCreatedFunction": ["dataLineCreatedFunctionWatcher"],
        "dataLines": ["dataLinesWatcher"],
        "excludedBarColor": ["excludedBarColorWatcher"],
        "includedBarColor": ["includedBarColorWatcher"],
        "labelFormatFunction": ["labelFormatFunctionWatcher"],
        "max": ["maxWatcher"],
        "min": ["minWatcher"],
        "precision": ["precisionWatcher"],
        "rangeType": ["rangeTypeWatcher"],
        "standardDeviation": ["standardDeviationWatcher"],
        "standardDeviationCount": ["standardDeviationCountWatcher"],
        "values": ["valuesWatcher"]
    }; }
    static get style() { return histogramRangeSliderCss; }
}, [0, "arcgis-histogram-range-slider", {
        "referenceElement": [1, "reference-element"],
        "average": [2],
        "barCreatedFunction": [16],
        "bins": [16],
        "dataLineCreatedFunction": [16],
        "dataLines": [16],
        "excludedBarColor": [16],
        "includedBarColor": [16],
        "labelFormatFunction": [16],
        "max": [2],
        "min": [2],
        "precision": [2],
        "rangeType": [1, "range-type"],
        "standardDeviation": [2, "standard-deviation"],
        "standardDeviationCount": [2, "standard-deviation-count"],
        "values": [16],
        "position": [1],
        "expandable": [4],
        "widget": [1040]
    }, undefined, {
        "average": ["averageWatcher"],
        "barCreatedFunction": ["barCreatedFunctionWatcher"],
        "bins": ["binsWatcher"],
        "dataLineCreatedFunction": ["dataLineCreatedFunctionWatcher"],
        "dataLines": ["dataLinesWatcher"],
        "excludedBarColor": ["excludedBarColorWatcher"],
        "includedBarColor": ["includedBarColorWatcher"],
        "labelFormatFunction": ["labelFormatFunctionWatcher"],
        "max": ["maxWatcher"],
        "min": ["minWatcher"],
        "precision": ["precisionWatcher"],
        "rangeType": ["rangeTypeWatcher"],
        "standardDeviation": ["standardDeviationWatcher"],
        "standardDeviationCount": ["standardDeviationCountWatcher"],
        "values": ["valuesWatcher"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-histogram-range-slider", "arcgis-expand"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-histogram-range-slider":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisHistogramRangeSlider$1);
            }
            break;
        case "arcgis-expand":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const ArcgisHistogramRangeSlider = ArcgisHistogramRangeSlider$1;
const defineCustomElement = defineCustomElement$1;

export { ArcgisHistogramRangeSlider, defineCustomElement };
