// @esri/calcite-components/dist/esm/da-d82a31f1.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
import { d as dayjs } from './calcite-input-time-picker.calcite-time-picker-c4d9b829.js';
import './index-2e6dbb80.js';
import './form-3f199679.js';
import './dom-bea5da81.js';
import './guid-9a252389.js';
import './resources-ed5aea48.js';
import './interactive-a992f5df.js';
import './browser-26f0727c.js';
import './key-c94e53b5.js';
import './label-ce4d8e83.js';
import './component-ec9f9062.js';
import './loadable-0092c6b2.js';
import './locale-ccbc05e8.js';
import './observers-fe652afa.js';
import './focusTrapComponent-6b2f2d88.js';
import './math-246b75fd.js';
import './t9n-e5ee6b09.js';
import './openCloseComponent-fb5847f4.js';
import './Validation-3bdaf762.js';

// Danish [da]
var locale = {
  name: 'da',
  weekdays: 'søndag_mandag_tirsdag_onsdag_torsdag_fredag_lørdag'.split('_'),
  weekdaysShort: 'søn._man._tirs._ons._tors._fre._lør.'.split('_'),
  weekdaysMin: 'sø._ma._ti._on._to._fr._lø.'.split('_'),
  months: 'januar_februar_marts_april_maj_juni_juli_august_september_oktober_november_december'.split('_'),
  monthsShort: 'jan._feb._mar._apr._maj_juni_juli_aug._sept._okt._nov._dec.'.split('_'),
  weekStart: 1,
  ordinal: function ordinal(n) {
    return n + ".";
  },
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD.MM.YYYY',
    LL: 'D. MMMM YYYY',
    LLL: 'D. MMMM YYYY HH:mm',
    LLLL: 'dddd [d.] D. MMMM YYYY [kl.] HH:mm'
  },
  relativeTime: {
    future: 'om %s',
    past: '%s siden',
    s: 'få sekunder',
    m: 'et minut',
    mm: '%d minutter',
    h: 'en time',
    hh: '%d timer',
    d: 'en dag',
    dd: '%d dage',
    M: 'en måned',
    MM: '%d måneder',
    y: 'et år',
    yy: '%d år'
  }
};
dayjs.locale(locale, null, true);

export default locale;
