// @esri/calcite-components/dist/esm/it-ch-786e144d.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
import { d as dayjs } from './calcite-input-time-picker.calcite-time-picker-c4d9b829.js';
import './index-2e6dbb80.js';
import './form-3f199679.js';
import './dom-bea5da81.js';
import './guid-9a252389.js';
import './resources-ed5aea48.js';
import './interactive-a992f5df.js';
import './browser-26f0727c.js';
import './key-c94e53b5.js';
import './label-ce4d8e83.js';
import './component-ec9f9062.js';
import './loadable-0092c6b2.js';
import './locale-ccbc05e8.js';
import './observers-fe652afa.js';
import './focusTrapComponent-6b2f2d88.js';
import './math-246b75fd.js';
import './t9n-e5ee6b09.js';
import './openCloseComponent-fb5847f4.js';
import './Validation-3bdaf762.js';

// Italian (Switzerland) [it-ch]
var locale = {
  name: 'it-ch',
  weekdays: 'domenica_lunedì_martedì_mercoledì_giovedì_venerdì_sabato'.split('_'),
  months: 'gennaio_febbraio_marzo_aprile_maggio_giugno_luglio_agosto_settembre_ottobre_novembre_dicembre'.split('_'),
  weekStart: 1,
  weekdaysShort: 'dom_lun_mar_mer_gio_ven_sab'.split('_'),
  monthsShort: 'gen_feb_mar_apr_mag_giu_lug_ago_set_ott_nov_dic'.split('_'),
  weekdaysMin: 'do_lu_ma_me_gi_ve_sa'.split('_'),
  ordinal: function ordinal(n) {
    return n;
  },
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD.MM.YYYY',
    LL: 'D MMMM YYYY',
    LLL: 'D MMMM YYYY HH:mm',
    LLLL: 'dddd D MMMM YYYY HH:mm'
  },
  relativeTime: {
    future: 'tra %s',
    past: '%s fa',
    s: 'alcuni secondi',
    m: 'un minuto',
    mm: '%d minuti',
    h: 'un\'ora',
    hh: '%d ore',
    d: 'un giorno',
    dd: '%d giorni',
    M: 'un mese',
    MM: '%d mesi',
    y: 'un anno',
    yy: '%d anni'
  }
};
dayjs.locale(locale, null, true);

export default locale;
