// @esri/calcite-components/dist/esm/pt-br-c0fd2736.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
import { d as dayjs } from './calcite-input-time-picker.calcite-time-picker-c4d9b829.js';
import './index-2e6dbb80.js';
import './form-3f199679.js';
import './dom-bea5da81.js';
import './guid-9a252389.js';
import './resources-ed5aea48.js';
import './interactive-a992f5df.js';
import './browser-26f0727c.js';
import './key-c94e53b5.js';
import './label-ce4d8e83.js';
import './component-ec9f9062.js';
import './loadable-0092c6b2.js';
import './locale-ccbc05e8.js';
import './observers-fe652afa.js';
import './focusTrapComponent-6b2f2d88.js';
import './math-246b75fd.js';
import './t9n-e5ee6b09.js';
import './openCloseComponent-fb5847f4.js';
import './Validation-3bdaf762.js';

// Portuguese (Brazil) [pt-br]
var locale = {
  name: 'pt-br',
  weekdays: 'domingo_segunda-feira_terça-feira_quarta-feira_quinta-feira_sexta-feira_sábado'.split('_'),
  weekdaysShort: 'dom_seg_ter_qua_qui_sex_sáb'.split('_'),
  weekdaysMin: 'Do_2ª_3ª_4ª_5ª_6ª_Sá'.split('_'),
  months: 'janeiro_fevereiro_março_abril_maio_junho_julho_agosto_setembro_outubro_novembro_dezembro'.split('_'),
  monthsShort: 'jan_fev_mar_abr_mai_jun_jul_ago_set_out_nov_dez'.split('_'),
  ordinal: function ordinal(n) {
    return n + "\xBA";
  },
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD/MM/YYYY',
    LL: 'D [de] MMMM [de] YYYY',
    LLL: 'D [de] MMMM [de] YYYY [às] HH:mm',
    LLLL: 'dddd, D [de] MMMM [de] YYYY [às] HH:mm'
  },
  relativeTime: {
    future: 'em %s',
    past: 'há %s',
    s: 'poucos segundos',
    m: 'um minuto',
    mm: '%d minutos',
    h: 'uma hora',
    hh: '%d horas',
    d: 'um dia',
    dd: '%d dias',
    M: 'um mês',
    MM: '%d meses',
    y: 'um ano',
    yy: '%d anos'
  }
};
dayjs.locale(locale, null, true);

export default locale;
