// @esri/calcite-components/dist/esm/zh-hk-5f93fc98.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
import { d as dayjs } from './calcite-input-time-picker.calcite-time-picker-c4d9b829.js';
import './index-2e6dbb80.js';
import './form-3f199679.js';
import './dom-bea5da81.js';
import './guid-9a252389.js';
import './resources-ed5aea48.js';
import './interactive-a992f5df.js';
import './browser-26f0727c.js';
import './key-c94e53b5.js';
import './label-ce4d8e83.js';
import './component-ec9f9062.js';
import './loadable-0092c6b2.js';
import './locale-ccbc05e8.js';
import './observers-fe652afa.js';
import './focusTrapComponent-6b2f2d88.js';
import './math-246b75fd.js';
import './t9n-e5ee6b09.js';
import './openCloseComponent-fb5847f4.js';
import './Validation-3bdaf762.js';

// Chinese (Hong Kong) [zh-hk]
var locale = {
  name: 'zh-hk',
  months: '一月_二月_三月_四月_五月_六月_七月_八月_九月_十月_十一月_十二月'.split('_'),
  monthsShort: '1月_2月_3月_4月_5月_6月_7月_8月_9月_10月_11月_12月'.split('_'),
  weekdays: '星期日_星期一_星期二_星期三_星期四_星期五_星期六'.split('_'),
  weekdaysShort: '週日_週一_週二_週三_週四_週五_週六'.split('_'),
  weekdaysMin: '日_一_二_三_四_五_六'.split('_'),
  ordinal: function ordinal(number, period) {
    switch (period) {
      case 'W':
        return number + "\u9031";

      default:
        return number + "\u65E5";
    }
  },
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'YYYY/MM/DD',
    LL: 'YYYY年M月D日',
    LLL: 'YYYY年M月D日 HH:mm',
    LLLL: 'YYYY年M月D日dddd HH:mm'
  },
  relativeTime: {
    future: '%s內',
    past: '%s前',
    s: '幾秒',
    m: '一分鐘',
    mm: '%d 分鐘',
    h: '一小時',
    hh: '%d 小時',
    d: '一天',
    dd: '%d 天',
    M: '一個月',
    MM: '%d 個月',
    y: '一年',
    yy: '%d 年'
  }
};
dayjs.locale(locale, null, true);

export default locale;
