// @esri/calcite-components/dist/cjs/it-ch-6c3c624e.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
'use strict';

const calciteInputTimePicker_calciteTimePicker_entry = require('./calcite-input-time-picker.calcite-time-picker-d8ae72cb.js');
require('./index-ac8ddc40.js');
require('./form-8075c2bf.js');
require('./dom-1c95ec06.js');
require('./guid-b54d4f6f.js');
require('./resources-1015acf8.js');
require('./interactive-8ea1ee84.js');
require('./browser-00f6569a.js');
require('./key-15fe7e2f.js');
require('./label-d786c0cb.js');
require('./component-894fc47a.js');
require('./loadable-96af681d.js');
require('./locale-a5749882.js');
require('./observers-0e539342.js');
require('./focusTrapComponent-781af75d.js');
require('./math-26a12122.js');
require('./t9n-9c5bc36c.js');
require('./openCloseComponent-fde95b54.js');
require('./Validation-27cda8b8.js');

// Italian (Switzerland) [it-ch]
var locale = {
  name: 'it-ch',
  weekdays: 'domenica_lunedì_martedì_mercoledì_giovedì_venerdì_sabato'.split('_'),
  months: 'gennaio_febbraio_marzo_aprile_maggio_giugno_luglio_agosto_settembre_ottobre_novembre_dicembre'.split('_'),
  weekStart: 1,
  weekdaysShort: 'dom_lun_mar_mer_gio_ven_sab'.split('_'),
  monthsShort: 'gen_feb_mar_apr_mag_giu_lug_ago_set_ott_nov_dic'.split('_'),
  weekdaysMin: 'do_lu_ma_me_gi_ve_sa'.split('_'),
  ordinal: function ordinal(n) {
    return n;
  },
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD.MM.YYYY',
    LL: 'D MMMM YYYY',
    LLL: 'D MMMM YYYY HH:mm',
    LLLL: 'dddd D MMMM YYYY HH:mm'
  },
  relativeTime: {
    future: 'tra %s',
    past: '%s fa',
    s: 'alcuni secondi',
    m: 'un minuto',
    mm: '%d minuti',
    h: 'un\'ora',
    hh: '%d ore',
    d: 'un giorno',
    dd: '%d giorni',
    M: 'un mese',
    MM: '%d mesi',
    y: 'un anno',
    yy: '%d anni'
  }
};
calciteInputTimePicker_calciteTimePicker_entry.dayjs.locale(locale, null, true);

exports.default = locale;
