// @esri/calcite-components/dist/collection/utils/openCloseComponent.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.6.0
 */
import { readTask } from "@stencil/core";
const componentToTransitionListeners = new WeakMap();
function transitionStart(event) {
    if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
        isOpen(this) ? this.onBeforeOpen() : this.onBeforeClose();
    }
}
function transitionEnd(event) {
    if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
        isOpen(this) ? this.onOpen() : this.onClose();
    }
}
function isOpen(component) {
    return "opened" in component ? component.opened : component.open;
}
function emitImmediately(component, nonOpenCloseComponent = false) {
    (nonOpenCloseComponent ? component[component.transitionProp] : isOpen(component))
        ? component.onBeforeOpen()
        : component.onBeforeClose();
    (nonOpenCloseComponent ? component[component.transitionProp] : isOpen(component))
        ? component.onOpen()
        : component.onClose();
}
/**
 * Helper to determine globally set transition duration on the given openTransitionProp, which is imported and set in the @Watch("open").
 * Used to emit (before)open/close events both for when the opacity transition is present and when there is none (transition-duration is set to 0).
 *
 * @example
 * import { onToggleOpenCloseComponent, OpenCloseComponent } from "../../utils/openCloseComponent";
 *
 * async componentWillLoad() {
 * // When component initially renders, if `open` was set we need to trigger on load as watcher doesn't fire.
 * if (this.open) {
 *    onToggleOpenCloseComponent(this);
 * }
 * @Watch("open")
 * async toggleModal(value: boolean): Promise<void> {
 *    onToggleOpenCloseComponent(this);
 * }
 *
 * @param component - OpenCloseComponent uses `open` prop to emit (before)open/close.
 * @param nonOpenCloseComponent - OpenCloseComponent uses `expanded` prop to emit (before)open/close.
 */
export function onToggleOpenCloseComponent(component, nonOpenCloseComponent = false) {
    readTask(() => {
        if (component.transitionEl) {
            const { transitionDuration: allDurations, transitionProperty: allProps } = getComputedStyle(component.transitionEl);
            const allTransitionDurationsArray = allDurations.split(",");
            const allTransitionPropsArray = allProps.split(",");
            const openTransitionPropIndex = allTransitionPropsArray.indexOf(component.openTransitionProp);
            const transitionDuration = allTransitionDurationsArray[openTransitionPropIndex] ??
                /* Safari will have a single transition value if multiple props share it,
                so we fall back to it if there's no matching prop duration */
                allTransitionDurationsArray[0];
            if (transitionDuration === "0s") {
                emitImmediately(component, nonOpenCloseComponent);
                return;
            }
            const fallbackTimeoutId = setTimeout(() => {
                component.transitionEl.removeEventListener("transitionstart", onStart);
                component.transitionEl.removeEventListener("transitionend", onEndOrCancel);
                component.transitionEl.removeEventListener("transitioncancel", onEndOrCancel);
                emitImmediately(component, nonOpenCloseComponent);
            }, parseFloat(transitionDuration) * 1000);
            component.transitionEl.addEventListener("transitionstart", onStart);
            component.transitionEl.addEventListener("transitionend", onEndOrCancel);
            component.transitionEl.addEventListener("transitioncancel", onEndOrCancel);
            function onStart(event) {
                if (event.propertyName === component.openTransitionProp && event.target === component.transitionEl) {
                    clearTimeout(fallbackTimeoutId);
                    component.transitionEl.removeEventListener("transitionstart", onStart);
                    (nonOpenCloseComponent ? component[component.transitionProp] : isOpen(component))
                        ? component.onBeforeOpen()
                        : component.onBeforeClose();
                }
            }
            function onEndOrCancel(event) {
                if (event.propertyName === component.openTransitionProp && event.target === component.transitionEl) {
                    (nonOpenCloseComponent ? component[component.transitionProp] : isOpen(component))
                        ? component.onOpen()
                        : component.onClose();
                    component.transitionEl.removeEventListener("transitionend", onEndOrCancel);
                    component.transitionEl.removeEventListener("transitioncancel", onEndOrCancel);
                }
            }
        }
    });
}
/**
 * Helper to keep track of transition listeners on setTransitionEl and connectedCallback on OpenCloseComponent components.
 *
 * For component which do not have open prop, use `onToggleOpenCloseComponent` implementation.
 *
 * @param component
 * @deprecated Call `onToggleOpenClose` in `componentWillLoad` and `open` property watchers instead.
 */
export function connectOpenCloseComponent(component) {
    disconnectOpenCloseComponent(component);
    if (component.transitionEl) {
        const boundOnTransitionStart = transitionStart.bind(component);
        const boundOnTransitionEnd = transitionEnd.bind(component);
        componentToTransitionListeners.set(component, [
            component.transitionEl,
            boundOnTransitionStart,
            boundOnTransitionEnd,
        ]);
        component.transitionEl.addEventListener("transitionstart", boundOnTransitionStart);
        component.transitionEl.addEventListener("transitionend", boundOnTransitionEnd);
    }
}
/**
 * Helper to tear down transition listeners on disconnectedCallback on OpenCloseComponent components.
 *
 * @param component
 * @deprecated Call `onToggleOpenClose` in `componentWillLoad` and `open` property watchers instead.
 */
export function disconnectOpenCloseComponent(component) {
    if (!componentToTransitionListeners.has(component)) {
        return;
    }
    const [transitionEl, start, end] = componentToTransitionListeners.get(component);
    transitionEl.removeEventListener("transitionstart", start);
    transitionEl.removeEventListener("transitionend", end);
    componentToTransitionListeners.delete(component);
}
